""" bootstrap BaseRecalibration without a reference set of SNPs """ 

rule filter_recalibrated_variants_DP:
  input:
    "{data}/vcf/recalibrated/merged_recalibrated_2x_snps.vcf.gz"
  output:
    "{data}/vcf/recalibrated/merged_recalibrated_hardfiltered_2x_snps.vcf.gz"
  params:
    job_name = "filter_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/recalibrated/logs/vcf_recalibrated_variants_2xfilter.txt"
  message:
    "filter second pass bootstrap vcf calls  "
  resources: all_threads = 1 
  shell:
    """
      java -Xmx16g \
        -XX:+UseSerialGC \
        -jar {GATK} \
        -T VariantFiltration \
        -R {GENOME} \
        -V {input} \
        -filterName FS_QD_DP_filter \
        --filterExpression "DP < 20" \
        -o {output}
    """

rule get_only_snps:
  input:
    "{data}/vcf/recalibrated/merged_recalibrated_2xpass.vcf.gz"
  output:
    "{data}/vcf/recalibrated/merged_recalibrated_2x_snps.vcf.gz"
  params:
    job_name = "get_snps",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/recalibrated/logs/vcf_recalibrated_variants_snps.txt"
  message:
    "extract only snps"
  resources: all_threads = 1 
  shell:
    """
      java -Xmx16g \
        -XX:+UseSerialGC \
        -jar {GATK} \
        -T SelectVariants \
        -R {GENOME} \
        -V {input} \
        -selectType SNP \
        -o {output}
    """

rule merge_calibrated_vcfs_round_2:
  input:
    expand("{data}/vcf/recalibrated/filtered/{fastq}_hardfiltered_2x.vcf.gz",
    data=DATA, fastq=FASTQ)
  output:
    "{data}/vcf/recalibrated/merged_recalibrated_2xpass.vcf.gz"
  params:
    job_name = "merge_vcfs",
    memory = "select[mem>40] rusage[mem=40]",
    vcf_list = "{data}/vcf/recalibrated/vcf_list_to_merge_2x.list"
  log:
    "{data}/vcf/recalibrated/logs/vcf_recalibrated_merge_2x.txt"
  message:
    "merge third pass bootstrap vcf calls  "
  resources: all_threads = 1 
  shell:
    """
      for vcf in {input};
        do echo $vcf >> {params.vcf_list}
      done

      java -Xmx16g \
        -XX:+UseSerialGC \
        -jar {GATK} \
        -T CombineVariants \
        -R {GENOME} \
        -V {params.vcf_list} \
        -o {output}

    """

rule filter_secondpass_recalibrated_variants:
  input:
    "{data}/vcf/recalibrated/{fastq}_thirdbootstrap.vcf.gz"
  output:
    "{data}/vcf/recalibrated/filtered/{fastq}_hardfiltered_2x.vcf.gz"
  params:
    tmp = "{data}/vcf/recalibrated/filtered/{fastq}_hardfiltered_2x.vcf",
    job_name = "filter_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/recalibrated/logs/vcf_recalibrated_variants_{fastq}_2x.txt"
  message:
    "filter third pass bootstrap vcf calls  "
  resources: all_threads = 1 
  shell:
    """
    module load samtools
    module load htslib
    tabix -f -p vcf {input}
    
    vcffilter -f "FS < 30.0 & QD > 2.0" \
    {input} > {params.tmp}
    bgzip {params.tmp}
    tabix -p vcf {output}
    """

rule generate_third_round_vcfs:
  input:
    bam = "{data}/star/{fastq}_split.bam",
    bqsr = "{data}/vcf/recalibrated/{fastq}_recal_2x.table"
  output:
    "{data}/vcf/recalibrated/{fastq}_thirdbootstrap.vcf.gz"
  params:
    job_name = "boostrap_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/recalibrated/logs/{fastq}_thirdbootstrap.txt"
  message:
    "third pass bootstrap vcf calls  "
  resources: all_threads = 4 
  threads:
    4
  shell:
    """
      java -Xmx12g -XX:+UseSerialGC \
        -jar {GATK} \
        -T HaplotypeCaller \
        -nct {threads} \
        -R {GENOME} \
        -I {input.bam} \
        -BQSR {input.bqsr} \
        -dontUseSoftClippedBases \
        -stand_call_conf 20.0 \
        -o {output}
    """

rule base_recalibration_round_two:
  input:
    known_sites =
    "{data}/vcf/recalibrated/merged_recalibrated_hardfiltered_1x.vcf.gz",
    bam = "{data}/star/{fastq}_split.bam"
  output:
    "{data}/vcf/recalibrated/{fastq}_recal_2x.table"
  params:
    job_name = "recal_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/recalibrated/logs/{fastq}_recal_2x.txt"
  message:
    "second pass bootstrap vcf recalibration  "
  resources: all_threads = 4 
  threads:
    4
  shell:
    """
      java -Xmx16g \
        -XX:+UseSerialGC \
        -jar {GATK} \
        -T BaseRecalibrator \
        -nct {threads}  \
        -R {GENOME} \
        -I {input.bam} \
        --knownSites {input.known_sites} \
        -o {output}
    """

rule merge_calibrated_vcfs_round_1:
  input:
    expand("{data}/vcf/recalibrated/filtered/{fastq}_hardfiltered_1x.vcf.gz",
    data=DATA, fastq=FASTQ)
  output:
    "{data}/vcf/recalibrated/merged_recalibrated_hardfiltered_1x.vcf.gz"
  params:
    job_name = "merge_vcfs",
    memory = "select[mem>40] rusage[mem=40]",
    vcf_list = "{data}/vcf/recalibrated/vcf_list_to_merge_1x.list"
  log:
    "{data}/vcf/recalibrated/logs/vcf_recalibrated_merge_1x.txt"
  message:
    "merge second pass bootstrap vcf calls  "
  resources: all_threads = 1 
  shell:
    """
      for vcf in {input};
        do echo $vcf >> {params.vcf_list}
      done

      java -Xmx16g \
        -XX:+UseSerialGC \
        -jar {GATK} \
        -T CombineVariants \
        -R {GENOME} \
        -V {params.vcf_list} \
        -o {output}

    """

rule filter_firstpass_recalibrated_variants:
  input:
    "{data}/vcf/recalibrated/{fastq}_secondbootstrap.vcf.gz",
  output:
    "{data}/vcf/recalibrated/filtered/{fastq}_hardfiltered_1x.vcf.gz"
  params:
    tmp = "{data}/vcf/recalibrated/filtered/{fastq}_hardfiltered_1x.vcf",
    job_name = "filter_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/recalibrated/logs/vcf_recalibrated_variants_{fastq}.txt"
  message:
    "filter second pass bootstrap vcf calls  "
  resources: all_threads = 1 
  shell:
    """
    module load samtools
    module load htslib
    tabix -f -p vcf {input}
    
    vcffilter -f "FS < 30.0 & QD > 2.0" \
    {input} > {params.tmp}
    bgzip {params.tmp}
    tabix -p vcf {output}
    """

rule generate_second_round_vcfs:
  input:
    bam = "{data}/star/{fastq}_split.bam",
    bqsr = "{data}/vcf/recalibrated/{fastq}_recal_1x.table"
  output:
    "{data}/vcf/recalibrated/{fastq}_secondbootstrap.vcf.gz"
  params:
    job_name = "boostrap_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/recalibrated/logs/{fastq}_secondbootstrap.txt"
  message:
    "second pass bootstrap vcf calls  "
  resources: all_threads = 4 
  threads:
    4
  shell:
    """
      java -Xmx12g -XX:+UseSerialGC \
        -jar {GATK} \
        -T HaplotypeCaller \
        -nct {threads} \
        -R {GENOME} \
        -I {input.bam} \
        -BQSR {input.bqsr} \
        -dontUseSoftClippedBases \
        -stand_call_conf 20.0 \
        -o {output}
    """

rule base_recalibration_round_one:
  input:
    known_sites =
    "{data}/vcf/uncalibrated/merged_uncalibrated_filtered.vcf.gz",
    bam = "{data}/star/{fastq}_split.bam"
  output:
    "{data}/vcf/recalibrated/{fastq}_recal_1x.table"
  params:
    job_name = "recal_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/recalibrated/logs/{fastq}_recal_1x.txt"
  message:
    "first pass bootstrap vcf recalibration  "
  resources: all_threads = 4 
  threads:
    4
  shell:
    """
      java -Xmx16g \
        -XX:+UseSerialGC \
        -jar {GATK} \
        -T BaseRecalibrator \
        -nct {threads}  \
        -R {GENOME} \
        -I {input.bam} \
        --knownSites {input.known_sites} \
        -o {output}
    """

rule merge_uncalibrated_vcfs:
  input:
    expand("{data}/vcf/uncalibrated/filtered/{fastq}_hardfiltered.vcf.gz",
    data=DATA, fastq=FASTQ)
  output:
    "{data}/vcf/uncalibrated/merged_uncalibrated_filtered.vcf.gz"
  params:
    job_name = "merge_vcfs",
    memory = "select[mem>50] rusage[mem=50]",
    vcf_list = "{data}/vcf/uncalibrated/vcf_list_to_merge.list"
  log:
    "{data}/vcf/uncalibrated/logs/vcf_uncalibrated_merge.txt"
  message:
    "merge first pass bootstrap vcf calls  "
  resources: all_threads = 1 
  shell:
    """
      cat > {params.vcf_list}
      for vcf in {input};
        do echo $vcf >> {params.vcf_list}
      done

      java -Xmx16g \
        -XX:+UseSerialGC \
        -jar {GATK} \
        -T CombineVariants \
        -R {GENOME} \
        -V {params.vcf_list} \
        -o {output}

    """

rule filter_uncalibrated_variants:
  input:
    "{data}/vcf/uncalibrated/{fastq}_firstbootstrap.vcf.gz"
  output:
    "{data}/vcf/uncalibrated/filtered/{fastq}_hardfiltered.vcf.gz"
  params:
    tmp = "{data}/vcf/uncalibrated/filtered/{fastq}_hardfiltered.vcf",
    job_name = "filter_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/uncalibrated/logs/vcf_uncalibrated_variants_{fastq}.txt"
  message:
    "filter first pass bootstrap vcf calls  "
  resources: all_threads = 1 
  shell:
    """
    module load samtools # tabix and bgzip not in samtools/1.5 dir
    module load htslib
    tabix -f -p vcf {input}

    vcffilter -f "FS < 30.0 & QD > 2.0" \
    {input} > {params.tmp}
    bgzip {params.tmp}
    tabix -p vcf {output}
    """

rule generate_uncalibrated_snp_db:
  input:
    "{data}/star/{fastq}_split.bam"
  output:
    "{data}/vcf/uncalibrated/{fastq}_firstbootstrap.vcf.gz"
  params:
    job_name = "boostrap_vcfs",
    memory = "select[mem>25] rusage[mem=25]",
  log:
    "{data}/vcf/uncalibrated/logs/{fastq}_firstbootstrap.txt"
  message:
    "first pass bootstrap vcf calls  "
  resources: all_threads = 4 
  threads:
    4 
  shell:
    """
      java -Xmx12g -XX:+UseSerialGC \
        -jar {GATK} \
        -T HaplotypeCaller \
        -nct {threads} \
        -R {GENOME} \
        -I {input} \
        -dontUseSoftClippedBases \
        -stand_call_conf 20.0 \
        -o {output}
    """
