
rule run_reports:
  input:
    "{data}/vcf/{variant_types}/{pos_ref}_{pos_alt}_alleles/filtered_select_variants.bed.gz",
    counts = expand("{{data}}/vcf/{{variant_types}}/{{pos_ref}}_{{pos_alt}}_alleles/counts/{fastq}_counts.txt.gz",
    fastq=FASTQ)
  output:
    "{data}/../results/{result_date}/{variant_types}/{pos_ref}_{pos_alt}_alleles/{pos_ref}_{pos_alt}_alleles.html"
  params:
    allele = "{pos_ref}_{pos_alt}_alleles",
    indir = "{data}/vcf/{variant_types}/",
    outdir = os.path.join(RESULTS, "{result_date}", "{variant_types}"),
    job_name = "run_reports",
    memory = "select[mem>65] rusage[mem=65]",
    rmd = os.path.join(LIB, "analyze_all_variants.Rmd"),
    pandoc = PANDOC
  resources: all_threads=2
  log:
    os.path.join(RESULTS, "logs/{result_date}/{variant_types}/run_all_variants_{pos_ref}_{pos_alt}.txt")
  shell:
    """
    Rscript {LIB}/render_single_variants.R \
      {params.allele} \
      {params.indir} \
      {params.outdir} \
      {params.rmd} \
      {params.outdir} \
      {params.pandoc}
    """


