---
title: "generate input for giremi"
author: "Kent Riemondy RBI"
date: "9/18/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Prepare input for Giremi

### select only sites called heterozygous

```{r get_hets}
vcfs <- list.files(file.path(data_dir, "vcf/variant_allele_counts_by_strand/"), 
                   pattern = "filtered_select_variants.vcf.gz$",
                   recursive = T,
                   full.names = T)

vcfs <- str_subset(vcfs, "_alleles")

vcf_dfs <- purrr::map(vcfs, ~read_tsv(.x, comment = "##"))
names_vcf <- str_match(vcfs, "[ATCG]_[ATCG]_alleles") 

names(vcf_dfs) <- names_vcf

# make site annotation col and drop other cols
vcf_dfs <- map(vcf_dfs, ~mutate(.x, site = paste(`#CHROM`, POS, REF, sep = "::")))

# count the number of het and homozygous variant calls per site
map(vcf_dfs, ~.x %>% 
  mutate_at(.vars = 10:ncol(.), 
            .funs = funs(hets = str_detect(., "0/1"), 
                         homozy = str_detect(., "1/1"))) %>% 
  mutate(hets = rowSums(.[, str_which(colnames(.), "hets")]),
         homozy = rowSums(.[, str_which(colnames(.), "homozy")],),
         geneid = str_match(INFO, "DENOVO_GENE_ID=([^;]+)")[, 2]) %>% 
  dplyr::select(geneid, site, hets, homozy)) -> het_homozy_calls


het_homozy_calls <- bind_rows(het_homozy_calls, .id = "allele")

het_homozy_calls <- mutate(het_homozy_calls, 
                           total_variant_calls = hets + homozy,
                           geneid = ifelse(is.na(geneid),
                                           "Inte",
                                           geneid))

# only keep het sites with at least 3 het variant calls
het_homozy_calls %>% 
  dplyr::filter(hets >= 3) -> potential_hets
```

```{r}
files <- list.files("edits/variant_allele_counts_by_strand/",
                    pattern = "filtered_sites_raw_counts.txt.gz$",
                    full.names = T,
                    recursive = T
                    )

# exclude A-to-G variants for SNV dataset
#files <- files[!str_detect(files, "A_G")]

dat <- map(files, read_tsv)

names(dat) <- basename(files)

dat <- map(dat, 
           ~dplyr::select(.x, 
                          edit_site))

dat <- bind_rows(dat, .id = "filename")

dat <- dat %>% 
  separate(edit_site, c("chrom", "pos", "allele"), 
           sep = "::", remove = F)

dat <- dat %>% 
  mutate(reference = str_extract(filename, "^[ATCG]"),
         strand = ifelse(reference == allele, 
                         "+",
                         "-"))
bed_dat <- dat %>% 
  dplyr::mutate(start = as.integer(pos) - 1L, 
                end = pos,
                score = ifelse(str_detect(filename, "A_G"),
                               0,
                               1)) %>% 
  dplyr::select(chrom, start, end, edit_site, score, strand) %>% 
  unique()
     
het_bed <- inner_join(bed_dat,
                     potential_hets,
                     by = c("edit_site" = "site"))

het_bed <- dplyr::select(het_bed,
                         chrom:end, 
                         geneid, 
                         score,
                         strand)
write_tsv(het_bed, 
          "all_filtered_het_variants.bed",
          col_names = F)             

```
