---
title: "edits-anova-vs-glm"
author: "Kent Riemondy RBI"
date: "6/25/2018"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    fig_caption: yes
    fig_retina: 1 
    code_folding: hide
params:
  allele_counts: "/Users/kriemo/Projects/Martin/data/vcf/variant_allele_counts_by_strand/A_G_alleles/counts/"
  variant_info: "/Users/kriemo/Projects/Martin/data/vcf/variant_allele_counts_by_strand/A_G_alleles/filtered_select_variants.bed.gz"
  ref: "A"
  alt: "G"
  prefix: "edits"
  dirname: "A_G_alleles"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = FALSE)
```

```{r ,message=FALSE, warning=FALSE, echo=FALSE}
library(dplyr)
library(purrr)
library(tidyr)
library(readr)
library(stringr)
library(ggplot2)
library(RColorBrewer)
library(edgeR)
library(ComplexHeatmap)
library(VennDiagram)
library(gridExtra)
library(cowplot)
```

### Differential editing analysis

```{r general_vars}

#colors mapping to state
state_cols <-  c(
  SA = rgb(255, 0, 0, maxColorValue=255),
  IBA = rgb(67, 205, 128, maxColorValue=255),
  Ent = rgb(155, 48, 255, maxColorValue=255),
  LT = rgb(25, 25, 112, maxColorValue=255),
  Ar = rgb(0, 0, 255, maxColorValue=255),
  SpD = rgb(255, 165, 0, maxColorValue=255) 
)
# preferred order
state_order = c(
  "SA", "IBA", "Ent", "LT", "Ar", "SpD"
)

#colors for region
region_cols <- c(
  Medulla = "#4DAF4A",
  Hypothalamus = "#377EB8",
  Forebrain =  "#E41A1C",
  Cerebrum = "#E41A1C"
)
# region order
region_order <- c("Medulla", "Hypothalamus", "Forebrain")

allele_counts_path <- params$allele_counts
ref <- params$ref
alt <- params$alt
variant_info <- params$variant_info
outname <- params$dirname
```

```{r io_util_functions}
bp_comp <- function(nt){
  #return complementary basepair
  comp <- c("A", "G", "T", "C")
  names(comp) <- c("T", "C", "A", "G")
  comp[nt] %>% unname()
}

group_dat <- function(.dir, .pattern){
  # read in and aggregate data into matrix-like format
  files <- dir(.dir, 
             full.names = T,
             pattern = .pattern)
  dat <- suppressMessages(map(files,  ~readr::read_tsv(.x,  col_names = T, progress = F)))
  dat <- Reduce(function(x, y) inner_join(x, y, by = c("#chrom", "pos", "ref", "base")), dat)
  colnames(dat) <- c("#chrom", "pos", "ref", "base", basename(files))
  dat
}  

write_gztsv <- function(df, name, ...){
  # write output as gzipped, supply name without .gz
  if(str_detect(name, ".gz$")){
    uncompressed_name <- str_replace(name, ".gz$", "")
  } else {
    uncompressed_name <- name
  }
  write_tsv(df, uncompressed_name, ...)
  system(paste0("gzip -f ", uncompressed_name))
}
```

### load in the data

```{r load_in_data}
ref_comp <- bp_comp(ref)
alt_comp <- bp_comp(alt)
# for some unknown reason this previous definition is not recognized when running on tesla. 

allele_counts_path <- params$allele_counts
prefix <- params$prefix
outname <- params$dirname
message(prefix)
message(allele_counts_path)
message(params$allele_counts)
message(params$dirname)

regions = c("medulla", "hypothalamus", "brainrest")
dat <- map(regions, ~group_dat(paste0(allele_counts_path, .x), ".gz"))
names(dat) <- regions

bind_cols <- c('#chrom', "pos", "ref", "base")
all_variants_export <- Reduce(function(x, y) inner_join(x, y, by = bind_cols), dat)
```

### clean input

```{r clean_up}

clean_and_filter <- function(dat, ref_nt, alt_nt, ref_comp, alt_comp){
  # filter for certain alleles, and spread each ref and alt count into seperate columns
  
  dat <- mutate(dat, 
                 edit_site = paste(`#chrom`, pos, ref, sep = "::")) %>% 
  dplyr::select(edit_site, everything())
  
  ag_dat <- filter(dat,
                   (ref == ref_nt & base == alt_nt) |
                   (ref == ref_nt & base == ref_nt) |
                   (ref == ref_comp & base == alt_comp) |
                  (ref == ref_comp & base == ref_comp)) %>%
    mutate(
           ref = ifelse(ref == ref_comp, ref_nt, ref),
           base = ifelse(base == ref_comp, ref_nt, 
                         ifelse(base == alt_comp, alt_nt, base)))
  
  ag_dat <- group_by(ag_dat, edit_site, `#chrom`, pos, ref)
  #make sure alternate allele is second in group
  if (ref_nt < alt_nt){
    ag_dat <- arrange(ag_dat, `#chrom`, pos, ref, base)
  } else {
    ag_dat <- arrange(ag_dat, `#chrom`, pos, ref, desc(base))
  }

  ag_dat
}

get_long_data <- function(dat, ref_nt, alt_nt, output_prefix){
  ag_dat <- group_by(dat, edit_site, `#chrom`, pos, ref)
  
  ag_wide <- summarize_at(ag_dat, vars(matches(".gz")),
               funs(a = nth(., 1),b = nth(., 2)))
  
  ag_wide <- ungroup(ag_wide) %>%
      dplyr::select(-`#chrom`, -pos, -ref) %>% 
      dplyr::select(edit_site, everything())
 
  ag_wide
}

get_proportion_dat <- function(dat, ref_nt, alt_nt, output_prefix){
  ag_dat <- group_by(dat, edit_site, `#chrom`, pos, ref)
  
  ag_wide <- summarize_at(ag_dat, vars(matches(".gz")),
               funs(nth(., 2) / (nth(., 1) + nth(., 2))))
  
  ag_wide <- ungroup(ag_wide) %>%
      dplyr::select(-`#chrom`, -pos, -ref) %>% 
      dplyr::select(edit_site, everything())
  
  ag_wide
}

get_matrix <- function(x) {
  ag_matrix <- as.data.frame(x)
  rownames(ag_matrix) <- ag_matrix[, 1]
  ag_matrix[, 1] <- NULL
  ag_matrix
}

filter_variants <- function(df){
  #take proportion data and select varaints with max > 0.05 and min < 0.95
  df %>% 
    gather(library, props, -edit_site) %>% 
    group_by(edit_site) %>% 
    summarize(smax = max(props, na.rm = T), smin = min(props, na.rm = T)) %>% 
    (.$smax > 0.05)
}

dat <- map(dat, ~clean_and_filter(.x, ref, alt, ref_comp, alt_comp) ) 
long_dat <- map2(dat, names(dat), ~get_long_data(.x, ref, alt, .y) ) 
prop_dat <- map2(dat, names(dat), ~get_proportion_dat(.x, ref, alt, .y) ) 

matrix_dat <- map(long_dat, ~get_matrix(.x))
```

```{r save_objs}
#save matricies for all sites
all_prop <- bind_cols(prop_dat)
all_prop <- all_prop[, c(TRUE, !(colnames(all_prop)[2:ncol(all_prop)] == 'edit_site'))] 

all_long <- bind_cols(long_dat)
all_long <- all_long[, c(TRUE, !(colnames(all_long)[2:ncol(all_long)] == 'edit_site'))] 

all_dat_export <- bind_cols(dat)
keep_cols <- c('edit_site', '#chrom', "pos", "ref", "base")
all_dat_export <- all_dat_export[, c(rep(TRUE, 5), !(colnames(all_dat_export)[6:ncol(all_dat_export)] %in% keep_cols))] 
```


```{r filter_variants}
variants_to_keep <- function(df){
  #take proportion data and select varaints with max > 0.05 and min < 0.95
  df %>% 
    gather(library, props, -edit_site) %>% 
    group_by(edit_site) %>% 
    summarize(smax = max(props, na.rm = T), 
              smin = min(props, na.rm = T)) -> min_max
  sites_to_keep <- min_max$smax > 0.05 & min_max$smin < 0.95
  min_max <- min_max[sites_to_keep, ]
  min_max["edit_site"]
}

good_variants <- variants_to_keep(all_prop)

filtered_prop <- all_prop %>% semi_join(good_variants, by = "edit_site")
filtered_dat_export <- all_dat_export %>% semi_join(good_variants, by = "edit_site")
filtered_long <- all_long %>% semi_join(good_variants, by = "edit_site")


long_dat <- map(long_dat,  ~.x %>% semi_join(good_variants, by = "edit_site"))
prop_dat <- map(prop_dat,  ~.x %>% semi_join(good_variants, by = "edit_site"))
matrix_dat <- map(matrix_dat,  ~.x[rownames(.x) %in% good_variants$edit_site, ])

knitr::kable(data_frame(original_sites = length(all_prop$edit_site), 
                        after_filtering = length(filtered_prop$edit_site)))
```



```{r}
get_groups <- function(matrix_dat, regex_for_fastq, ref_nt, alt_nt, pdata_file = "~/Projects/Martin/docs/BrainRegionRNAseqLibMetadata.txt") {
 # make_groups
  g_names <- colnames(matrix_dat)
  animal_number <- str_extract(g_names, regex_for_fastq)
  
  allele_type <- rep(c(ref_nt, alt_nt), each = 30)
  groups <- data_frame(g_names, animal_number, allele_type)
  
  pdata <- readr::read_tsv(pdata_file)
  
  pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)
  
  pdata <- mutate(pdata, 
                  abbrev_sample = str_split(sample, "_", simplify = T)[, 1])
  
  pdata <- inner_join(groups, 
                      pdata, 
                      by = c("animal_number" = "abbrev_sample")) %>%
    dplyr::select(g_names, animal_number, allele_type, State)
  pdata
}

regexes <- paste0(c("M", "H", "B"), "[0-9]+")

pdata_groups <- map2(matrix_dat, regexes, ~get_groups(.x, .y, ref, alt))
pdata_groups_original <- pdata_groups 

.dir <- "../../../data/featurecounts/"
pattern <- "_summary.tsv.summary$"

lib_sizes <- read_tsv(paste0(.dir, "count_summary.tsv.summary"))
lib_sizes <- filter(lib_sizes, Status == "Assigned") %>% 
  gather(library, counts, -Status) %>% 
  mutate(library = basename(library) %>% 
           str_replace(., "_2pass_Aligned.sortedByCoord.out.bam", ""))

extract_lib_sizes <- function(matrix_dat, lib_sizes_df) {
  #get original mapped to exons count for selected libraries
  .names <- colnames(matrix_dat)
  .names <- str_replace(.names, "_counts.txt.gz_a$", "")
  .names <- str_replace(.names, "_counts.txt.gz_b$", "")
  .names <- data_frame(original_name = .names)
  sizes <- inner_join(.names, lib_sizes_df, by = c("original_name" = "library"))
  res <- dplyr::select(sizes, counts) %>% unlist()
  names(res) <- sizes[["library"]]
  res
}

lib_sizes <- map(matrix_dat, ~extract_lib_sizes(.x, lib_sizes))
```

```{r build_design_matrices}

get_design_matrix <- function(pdata, ref_nt){
  #set up design matrix
  design <- model.matrix(~0 + animal_number + State:allele_type, data = pdata)
  design <- design[,!grepl(paste0("allele_type", ref_nt), colnames(design))]
  design
}

design_matrices <- map(pdata_groups, ~get_design_matrix(.x, ref))

```


```{r initialize_edger_objs}

get_edger_objs <- function(matrix_dat, pdata, lib_size){
  dge <- DGEList(matrix_dat, group = pdata$State, lib.size = lib_size)
  keep <- rowSums(getCounts(dge)>5, na.rm = T) >= 4
  dge <- dge[keep, , keep.lib.sizes=TRUE]
  dge
}

edger_objs <- pmap(list(matrix_dat, pdata_groups, lib_sizes),
                   function(x, y, z) get_edger_objs(x, y, z))

```

### set up anova

Use only sites filtered as above. 

```{r get_pdata}

get_groups <- function(fastq_id, regex_for_fastq, pdata_file = "~/Projects/Martin/docs/BrainRegionRNAseqLibMetadata.txt") {
 # make_groups
  animal_number <- str_extract(fastq_id, regex_for_fastq)
  
  groups <- data_frame(fastq_id, animal_number)
  
  pdata <- readr::read_tsv(pdata_file)
  
  pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)
  
  pdata <- mutate(pdata, 
                  abbrev_sample = str_split(sample, "_", simplify = T)[, 1])
  
  pdata <- inner_join(groups, 
                      pdata, 
                      by = c("animal_number" = "abbrev_sample")) %>%
    dplyr::select(fastq_id, animal_number, State)
  pdata
}

regexes <- paste0(c("M", "H", "B"), "[0-9]+")

pdata_groups <- map2(prop_dat, regexes, 
                     ~get_groups(colnames(.x)[2:length(colnames(.x))], 
                                 .y))

library(doParallel)

registerDoParallel(3)

objs <- pmap(list(
  prop_dat,
  pdata_groups,
  edger_objs), function(x, y, z) list(x, y, z))

get_anova_results <- function(x, y, z) {
       tmp_dat <- x %>% 
         filter(edit_site %in% rownames(z$counts)) %>% 
         gather(library, editing_freq, -edit_site)
       tmp_dat <- left_join(tmp_dat, 
                       y, by = c("library" = "fastq_id"))
       tmp_dat$State <- factor(tmp_dat$State, 
                               levels = names(state_cols))
       a_stats <- dplyr::group_by(tmp_dat, edit_site) %>% 
         do(broom::tidy(anova(lm(editing_freq ~ State, data = .))))
       a_stats
}

ris <- foreach(i = objs,
              .packages = c("broom", "tidyverse"),
              .export = c("find_molecules", "bampath"))  %dopar% {
              get_anova_results(i[[1]], i[[2]], i[[3]])}

anova_res <- map(ris, ~filter(ungroup(.x), term == "State") %>% 
       mutate(padj = p.adjust(p.value, method = "fdr"))) 

names(anova_res) <- names(prop_dat)

fdr_filtered_sites <- map(anova_res, ~filter(.x, padj < 0.01) %>% pull(edit_site))
```


### plot heatmaps
```{r get_variant_info}
get_variant_info <- function(variant_info_file, ref, alt) {

  variant_info <- read_tsv(variant_info_file,
                         col_names = F)
    colnames(variant_info) <- c("chrom", "start", "end", "REF", "ALT", "site", "strand", "EFF", "Disruption", "GeneName", "GeneID", "Nucleotide", "AminoAcid", "AllANNInfo", 'GENE_ID', 'BIOTYPE', 'TRANSCRIPT_ID', 'GENE_NAME', 'DENOVO_GENE_ID','DENOVO_TRANSCRIPT_ID', 'EXON_NUMBER', 'ANNOTATED')

  variant_info_genenames <- variant_info %>% 
    dplyr::select(chrom, site, GeneName, EFF, start) %>% 
    mutate(GeneName = ifelse(is.na(GeneName), chrom, GeneName)) %>% 
    mutate(gene_id_unique = paste(GeneName, EFF, start, sep = "::")) %>% 
    dplyr::select(site, gene_id_unique)
  variant_info_genenames
}

rename_sites_as_genes <- function(fdr_sites, variant_info_df){
      fdr_sig_sites <- map(fdr_sites, ~as_data_frame(.x) %>% 
      inner_join(., variant_info_df, by = c("value" = "site")) %>% 
      dplyr::select(gene_id_unique) %>% unlist())
}


gene_variant_info <- get_variant_info(variant_info, ref, alt)
fdr_sites_genes <- rename_sites_as_genes(fdr_filtered_sites, gene_variant_info)
```

```{r}
format_heatmap_matrix <- function(prop_data, na_value = NA) {
  
  ag_matrix <- as.data.frame(prop_data)
  rownames(ag_matrix) <- ag_matrix[, 1]
  ag_matrix[, 1] <- NULL
  
  # remove all NA rows
  remove_na <- apply(ag_matrix, 1, function(x) !all(is.na(x)))
  ag_matrix <- ag_matrix[remove_na, ]
   
  #remove all rows with SD = 0 (basically rows with all 1 or 0)
  ag_matrix <- ag_matrix[apply(ag_matrix, 1, function(x) sd(x, na.rm = T) != 0), ]

  #remove all rows with only 10 non-na values
  remove_too_many_na <- apply(ag_matrix, 1, function(x) !(sum(is.na(x)) > 25))
  ag_matrix <- ag_matrix[remove_too_many_na, ]
  
  #replace NaNs with NA
   ag_matrix[] <- lapply(ag_matrix, function(x){ 
    x[is.nan(x)] <- na_value 
    x 
  })
  ag_matrix
}  

heatmap_matrices <- map(prop_dat, ~format_heatmap_matrix(.x))

# bind all proprtion data into one matrix, but drop repeated 'edit_site' columns
all_prop_dat <- bind_cols(prop_dat)
all_prop_dat <- all_prop_dat[, c(TRUE, !(str_detect(colnames(all_prop_dat)[2:ncol(all_prop_dat)],
                                                    'edit_site')))] 

all_data_heatmap_matrix <- format_heatmap_matrix(all_prop_dat)
  
all_prop_dat_genes <- all_data_heatmap_matrix %>% 
  tibble::rownames_to_column("site") %>% 
  inner_join(., gene_variant_info, by = c("site")) %>% 
  tibble::column_to_rownames(., "gene_id_unique") %>% 
  dplyr::select(-site)

all_prop_dat_genes <- all_prop_dat_genes * 100

all_sites <- unlist(fdr_sites_genes) %>% unique()

```

```{r plot_hmaps}

names(pdata_groups_original) <- c("Medulla",
                     "Hypothalamus",
                     "Forebrain")

pdata_all <- bind_rows(pdata_groups_original, .id = "region") %>%  
  filter(grepl("_a$", g_names)) %>% 
  mutate(g_names = str_replace(g_names, "_a$", "")) %>% 
  dplyr::select(g_names, everything())

pdata_all <- mutate(pdata_all, 
                    region = ifelse(region == "Forebrain",
                                    "Cerebrum",
                                    region))

```

```{r}

all_sig_sites <- all_prop_dat_genes[rownames(all_prop_dat_genes) %in% all_sites, ]

n_sites = nrow(all_sig_sites)

pdata <- data_frame( "names" = colnames(all_prop_dat_genes)) %>%
    mutate(simple_names = str_extract(names, "[HBM][0-9]+")) %>%
    inner_join(. , pdata_all, by = c("simple_names" = "animal_number")) %>%
    dplyr::select(names, simple_names, State, region) %>% as.data.frame() %>% 
    arrange(State)

pdata <- mutate(pdata, 
                region = ifelse(region == "Forebrain",
                               "Cerebrum", region))
reordered_cols <- pdata$names %>% unlist()

all_sig_sites <- all_sig_sites[reordered_cols]

pdata <- dplyr::rename(pdata, 
                             "Sample Groups" = State,
                             "Brain Region" = region)

ha2 <- rowAnnotation(df = pdata[c("Sample Groups", 
                                          "Brain Region")], 
                                    col = list(
                                      "Brain Region" = region_cols,
                                        "Sample Groups" = state_cols
                                               ),
                         annotation_legend_param = list(
                          "Sample Groups" = list(
                                        at = state_order, 
                                        labels = state_order)),
                         gap = unit(2, "mm"),
                         show_annotation_name = TRUE,
                         annotation_name_side = "top",
                         annotation_name_offset = unit(1, "mm"))

ht1 <- Heatmap(t(all_sig_sites),
        show_row_names = FALSE,
        show_column_names = FALSE,
        cluster_rows = T,
        show_row_dend = T,
        show_column_dend = T,
         column_title = paste0("Significant ", ref, "-to-", alt, " variants using ANOVA n = ", n_sites),
        heatmap_legend_param = list(
          title = "Percent\nEditing (%)"))

draw(ha2 + ht1 , annotation_legend_side = "left", 
     gap = unit(1, "mm"), 
   heatmap_legend_side = "left", row_dend_side = "left")

pdf("edits/sig_editing_sites_clustered_just_anova.pdf", 
    useDingbats = F, width = 8, height = 5)

  draw(ha2 + ht1 , annotation_legend_side = "left", 
     gap = unit(1, "mm"), 
     padding = unit(c(2, 2, 25, 2), 'mm'),
   heatmap_legend_side = "left", row_dend_side = "left")
dev.off()


```

## compare ANOVA sites to edgeR sites

```{r}
gatk_glm_sites <- read_tsv("edits/A_G_filtered_fdr0.01_sites_annotation_kmeans.txt.gz") %>% 
  pull(site) %>% 
  unique(.)

gatk_anova_sites <- unique(unlist(fdr_filtered_sites))

library(eulerr)

all_sites <- unique(union(gatk_glm_sites,
      gatk_anova_sites))

lgl_df <- map(list(edger = gatk_glm_sites,
                   anova = gatk_anova_sites), 
              ~ all_sites %in% .x) %>% as.data.frame()
set.seed(1)
fit <- euler(lgl_df)


pdf("edits/anova_edger_overlap.pdf")
  plot(fit, quantities = T, fill = brewer.pal(3, "Set1"), fill_alpha = 0.7,
     labels = c("Sites found by edgeR", "Sites found by ANOVA" ),
     counts = list(cex = 1.5, col = "black"), cex = 1.5)
dev.off()
```

```{r}

out <- bind_rows(anova_res, .id = "region")
write_gztsv(out, "edits/anova_results.txt")
```