---
title: "RNA editing site visualizations"
author: "Kent Riemondy RBI"
date: "8/28/2017"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```

```{r, message=FALSE, warning=FALSE}
source("globals.R")
```

```{r}
cnames <- c("chrom", "start", "comment","ref", 
            "alt", "qual", "filter", "info", "allele")

# read in vcfs, drop genotypes
vcf <- read_tsv(file.path(data_dir, 
                          "vcf", 
                          "variant_allele_counts_by_strand", 
                          "A_G_alleles", 
                          "filtered.vcf.gz"), 
                comment = "#", 
                col_types = c("cicccnccc", 
                              rep("-", 90)), 
                col_names = cnames)


# adjust to zero based stars
vcf <- mutate(vcf, start = start - 1)

vcf <- mutate(vcf, EFF = str_match(info, "ANN=[AGTC]\\|([^|]+)\\|([^|]+)\\|") %>% 
                .[, 2], 
              IMPACT = str_match(info, "ANN=[AGTC]\\|([^|]+)\\|([^|]+)\\|") %>% 
                .[, 3]) %>% 
              dplyr::select(chrom, start, EFF, IMPACT, 
                     ref, alt, info, -comment, -qual, -filter)

```

```{r}
annots <- read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             "A_G_alleles", "A_G_filtered_sites_annotation.txt.gz"))

fdr <- filter(annots, FDR < 0.01)

# EFF column contains predictions based on ensembl85
fdr <- dplyr::rename(fdr, EFF_ensembl = EFF)

```

```{r}
fdr_all_sites <- inner_join(vcf, fdr, 
                            by = c("chrom", "start")) %>% dplyr::select(-info)
```

## Clustering RNA edits into classes {.tabset .tabset-fade .tabset-pills}

```{r}

# editing frequencies
dat <- read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             "A_G_alleles", "A_G_filtered_sites_proportions.txt.gz"))
dat <- as.data.frame(dat)
rownames(dat) <- dat[, 1]
dat <- dat[, -1]

# convert to percentages
dat <- dat * 100

```


```{r}
fdr_all_sites <- inner_join(vcf, fdr, 
                            by = c("chrom", "start")) %>% dplyr::select(-info)

fdr_all_sitenames <- unique(fdr_all_sites[["site"]])

# get matrix
all_sig_sites_matrix <- dat[rownames(dat) %in% fdr_all_sitenames, ]
all_sig_sites_matrix <- as.matrix(all_sig_sites_matrix)

# convert NaNs to 0 for clustering, but not for visualization
clustering_matrix <- all_sig_sites_matrix
clustering_matrix[is.nan(clustering_matrix)] <- 0
# center and scale
cluster_data <- t(scale(t(clustering_matrix), center = T, scale = T))
```


```{r elbow}
elbow_calc <- function(mat, k){
  # compute sum of squares for k = 1
  wss <- (nrow(mat)-1)*sum(apply(mat,2,var))
  for( i in 2:k) {
    clust <- flexclust::kcca(mat, i, 
                             control = list(initcent = "kmeanspp"))
    wss[i] <- info(clust, "distsum")
  }
  data_frame(k = 1:k, 
             WSS  = wss)
}

elbow_dat <- elbow_calc(cluster_data, 10)
p <- ggplot(elbow_dat, aes(k, WSS)) + 
  geom_point() +
  geom_line()+
  scale_x_continuous(breaks = pretty_breaks()) +
  scale_y_continuous(labels = scales::comma) +
  labs(y = "Within groups sum of squares",
       x = "Number of clusters",
       title = "GATK Sites")

save_plot("edits/wss_plot.pdf", p)
```


```{r}
# run kmeans with kmeans++ initialization (rather than traditional random assignment)
set.seed(20170719)
e_km <- flexclust::kcca(cluster_data, k = 2, 
                        control = list(initcent = "kmeanspp"))
km_clusters <- e_km@cluster

pdata_file <- file.path(docs_dir, "BrainRegionRNAseqLibMetadata.txt") 

 # make_groups
g_names <- colnames(dat)
animal_number <- str_extract(g_names, "[MHB][0-9]+")
  
 # allele_type <- rep(c("A", "G"), each = 30)
groups <- data_frame(g_names, animal_number)
  
pdata <- readr::read_tsv(pdata_file)
  
pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)
  
pdata <- mutate(pdata,
                region = ifelse(region == "Forebrain",
                                "Cerebrum",
                                region))
pdata <- mutate(pdata, 
                  abbrev_sample = str_split(sample, "_", simplify = T)[, 1])
  
pdata <- inner_join(groups, 
                      pdata, 
                      by = c("animal_number" = "abbrev_sample")) %>%
    dplyr::select(g_names, animal_number, State, region)

pdata_heatmap <- data_frame( "names" = colnames(all_sig_sites_matrix)) %>%
      inner_join(., pdata, by = c("names" = "g_names")) %>%
      dplyr::select(names, State, region) %>% as.data.frame() %>% 
      arrange(State)

reordered_cols <- pdata_heatmap$names %>% unlist()

all_sig_sites_matrix <- all_sig_sites_matrix[, reordered_cols]

ha2 <- HeatmapAnnotation(df = pdata_heatmap[, c('State', 'region')], 
                                      col = list(
                                        region = region_cols,
                                          State = state_cols
                                                 ),
                           annotation_legend_param = list(
                             State = list(title = "Hiberation\nState",
                                          at = state_order, labels = state_order)))
```

```{r na_handling}
format_heatmap_matrix <- function(mat, na_value = NA) {
  
  # remove all NA rows
  remove_na <- apply(mat, 1, function(x) !all(is.na(x)))
  mat <- mat[remove_na, ]
   
  #remove all rows with SD = 0 (basically rows with all 1 or 0)
  mat <- mat[apply(mat, 1, function(x) sd(x, na.rm = T) != 0), ]

  #remove all rows with only 10 non-na values
  remove_too_many_na <- apply(mat, 1, 
                              function(x) !(sum(is.na(x)) > 60))
  mat <- mat[remove_too_many_na, ]
  
  #replace NaNs with NA
   mat[] <- sapply(mat, function(x){ 
    x[is.nan(x)] <- na_value 
    x 
  })
  mat
}  
```


### Hierarchical Clustering

### Hierarchical Clustering (Split 1)

```{r, fig.width= 16, fig.height= 20}
ht <- Heatmap(format_heatmap_matrix(all_sig_sites_matrix),
      #        col = circlize::colorRamp2(c(0, 1), c("white", "red")),
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          split = 2,
          top_annotation = ha2,
           row_title = "Editing Sites",
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")
```

### Kmeans (2)

```{r, fig.width= 16, fig.height= 20}
fmat <- format_heatmap_matrix(all_sig_sites_matrix)
sites_no_excessive_na <- rownames(fmat)
km_clusters_no_na <- km_clusters[names(km_clusters) %in% rownames(fmat)]
nsites = nrow(km_clusters_no_na)

# split on classes, ordered within class by hclust, excluding sites with too many NAs to perform hclust
ht <- Heatmap(fmat,
      #        col = circlize::colorRamp2(c(0, 1), c("white", "red")),
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          split = km_clusters_no_na,
          top_annotation = ha2,
           row_title = "Editing Sites",
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")
pdf("edits/rna_editing_kmeans2_fdr0.01.pdf", useDingbats = F, width = 8, height = 8)
draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")
dev.off()
```

### Kmeans class 1 impactful sites

```{r, fig.width= 16, fig.height= 20}
km_sites_1 <- names(km_clusters[km_clusters == 1])
km_sites_2 <- names(km_clusters[km_clusters == 2])



impactful_vcf <- filter(vcf, IMPACT == "HIGH" | IMPACT == "MODERATE")
fdr_impact_sites <- inner_join(impactful_vcf, fdr, 
                            by = c("chrom", "start")) %>% dplyr::select(-info)

fdr_impact_sites %>% 
  mutate(site_id = paste(GENE_NAME, GENE_ID, 
                         EFF, start, sep = "::")) -> fdr_impact_sites

class1 <- fdr_impact_sites %>% 
  dplyr::filter(site %in% km_sites_1)

# convert matrix to rownames with unique site ids
gene_variant_info <- fdr_impact_sites %>% 
  dplyr::select(site, site_id) %>% unique()
all_prop_dat_genes <- dat %>% 
  tibble::rownames_to_column("site") %>% 
  inner_join(., gene_variant_info, by = c("site")) %>% 
  tibble::column_to_rownames(., "site_id") %>% 
  dplyr::select(-site)

#regenerate ha2 annoations with columns of all_prop_dat_genes
pdata_heatmap_genes <- data_frame( "names" = colnames(all_prop_dat_genes)) %>%
      inner_join(., pdata, by = c("names" = "g_names")) %>%
      dplyr::select(names, State, region) %>% as.data.frame() %>% 
      arrange(State)
  
reordered_cols <- pdata_heatmap_genes$names %>% unlist()

all_prop_dat_genes <- all_prop_dat_genes[, reordered_cols]

class1_sites <- all_prop_dat_genes[rownames(all_prop_dat_genes) %in% class1$site_id, ]

ha2 <- HeatmapAnnotation(df = pdata_heatmap_genes[c('State', 'region')], 
                                      col = list(
                                        region = region_cols,
                                          State = state_cols
                                                 ),
                           annotation_legend_param = list(
                             State = list(title = "Hiberation\nState",
                                          at = state_order, labels = state_order)))

ht <- Heatmap(class1_sites,
      #        col = circlize::colorRamp2(c(0, 1), c("white", "red")),
          show_row_names = T,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = F,
          cluster_columns = F,
          top_annotation = ha2,
       row_title = "Editing Sites",
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")
```

### Kmeans class 2 impactful sites

```{r, fig.width= 16, fig.height= 20}

class2 <- fdr_impact_sites %>% 
  filter(site %in% km_sites_2)

class2_sites <- all_prop_dat_genes[rownames(all_prop_dat_genes) %in% class2$site_id, ]

ht <- Heatmap(class2_sites,
      #        col = circlize::colorRamp2(c(0, 1), c("white", "red")),
          show_row_names = T,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = F,
          cluster_columns = F,
          top_annotation = ha2,
       row_title = "Editing Sites",
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")
```

### Rotated Kmeans 2 classes 

```{r, fig.width= 12, fig.height= 8}
# use t function and rowAnnotation to show sites as columns. Needs many modifications to work, but looks nice.
# in order to split by col, need to generate two heatmaps
km1 <- km_clusters_no_na[km_clusters_no_na == 1]
km2 <- km_clusters_no_na[km_clusters_no_na == 2]
fmat1 <- fmat[rownames(fmat) %in% names(km1), ]
fmat2 <- fmat[rownames(fmat) %in% names(km2), ]
fmat1 <- t(fmat1)
fmat2 <- t(fmat2)
n1 <- ncol(fmat1)
n2 <- ncol(fmat2)
nsites = ncol(fmat)

pdata_heatmap <- dplyr::rename(pdata_heatmap, 
                               "Sample Groups" = State,
                               "Brain Region" = region)

ha2 <- rowAnnotation(df = pdata_heatmap[c("Sample Groups", "Brain Region")], 
                                      col = list(
                                        "Brain Region" = region_cols,
                                          "Sample Groups" = state_cols
                                                 ),
                           annotation_legend_param = list(
                            "Sample Groups" = list(
                                          at = state_order, 
                                          labels = state_order)),
                           gap = unit(2, "mm"),
                           show_annotation_name = TRUE,
                           annotation_name_side = "top",
                           annotation_name_offset = unit(1, "mm"))

ht1 <- Heatmap(fmat1,
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          show_column_dend = F,
          #row_title = "Individual RNA-Seq libraries",
          #row_title_side = c("right"),
         column_title = "Group I Editing Sites",
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

ht2 <- Heatmap(fmat2,
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          show_column_dend = F,
          row_title = "Individual RNA-Seq libraries",
          row_title_side = c("right"),
          column_title = "Group 2",
          show_heatmap_legend = FALSE
         )

draw(ha2 + ht1 + ht2, annotation_legend_side = "left", gap = unit(1, "mm"), 
     heatmap_legend_side = "left", row_dend_side = "left")
pdf("edits/kmeans2_rotated_heatmap.pdf", useDingbats = F, width = 8, height = 5)
draw(ha2 + ht1 + ht2, annotation_legend_side = "left", gap = unit(1, "mm"), 
     heatmap_legend_side = "left", row_dend_side = "left")
dev.off()


```

```{r}
## make simpler heatmap for Emily W.'s RBI intern presentation
# drop region annotation and class II sites and label with larger simpler titles

ha2 <- rowAnnotation(df = pdata_heatmap[c("Sample Groups")], 
                                      col = list(
                                          "Sample Groups" = state_cols
                                                 ),
                           annotation_legend_param = list(
                            "Sample Groups" = list(
                                          at = state_order, 
                                          labels = state_order,
                                          title = "Hibernation\nState",
                                          title_gp = gpar(fontsize = 14), 
            labels_gp = gpar(fontsize = 12),
            legend_height = unit(2.5, "cm"))),
                           annotation_width = 8,
                           gap = unit(1, "mm"))

ht1 <- Heatmap(fmat1,
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          show_column_dend = F,
          row_title_side = c("right"),
          column_title = paste0("RNA Editing Sites (n = ", n1, ")"),
          row_title = "Individual RNA-Seq libraries",
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)",
            title_gp = gpar(fontsize = 14), 
            labels_gp = gpar(fontsize = 12))
         )

draw(ha2 + ht1 , 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left", 
     row_dend_side = "left",
     column_title_gp = gpar(fontsize = 20),
     row_title_gp = gpar(fontsize = 20))

pdf("edits/simplified_hmap_for_emily.pdf", useDingbats = F, width = 8, height = 5)
draw(ha2 + ht1 , 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left", 
     row_dend_side = "left",
     column_title_gp = gpar(fontsize = 20),
     row_title_gp = gpar(fontsize = 20))
dev.off()
```

### Rotated Kmeans 2 classes with dendo

```{r, fig.width= 12, fig.height= 8}

ha2 <- rowAnnotation(df = pdata_heatmap[c("Sample Groups", 
                                          "Brain Region")], 
                                      col = list(
                                        "Brain Region" = region_cols,
                                          "Sample Groups" = state_cols
                                                 ),
                           annotation_legend_param = list(
                            "Sample Groups" = list(
                                          at = state_order, 
                                          labels = state_order)),
                           gap = unit(2, "mm"),
                           show_annotation_name = TRUE,
                           annotation_name_side = "top",
                           annotation_name_offset = unit(1, "mm"))

ht1 <- Heatmap(fmat1,
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          show_column_dend = T,
          #row_title = "Individual RNA-Seq libraries",
          #row_title_side = c("right"),
          column_title = paste0("GATK\nGroup 1\n(n = ", formatC(n1, big.mark = ","), ")"), 
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

ht2 <- Heatmap(fmat2,
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          show_column_dend = T,
          row_title = "Individual RNA-Seq libraries",
          row_title_side = c("right"),
          column_title = paste0("GATK\nGroup 2\n(n = ", formatC(n2, big.mark = ","), ")"),
          show_heatmap_legend = FALSE
         )

draw(ha2 + ht1 + ht2, annotation_legend_side = "left", gap = unit(1, "mm"), 
     heatmap_legend_side = "left", row_dend_side = "left",
     padding = unit(c(4, 8, 4, 4), "mm"))

pdf("edits/kmeans2_rotated_heatmap_with_dendo.pdf", useDingbats = F, width = 8, height = 5)
draw(ha2 + ht1 + ht2, annotation_legend_side = "left", gap = unit(1, "mm"), 
     heatmap_legend_side = "left", row_dend_side = "left",
     padding = unit(c(4, 8, 4, 4), "mm"))
dev.off()

#fig 2b
file.copy("edits/kmeans2_rotated_heatmap_with_dendo.pdf", 
          figs_dir[2], overwrite = T)
```



### Only validation editing sites

```{r validation sites}

all_validation_sites <- c(
                   ZCCHC8 = "JH393368.1::2076972::A",
                   FBXW7 = "JH393499.1::2282179::T",
                   EIF3A = "JH393296.1::8517456::T",
                   ZC3H18 = "JH393451.1::1059495::T",
                   CHD9 = "JH393285.1::6065218::A",
                   ZNF483 = "JH393369.1::1779832::T",
                   AMIGO2 = "JH393322.1::5158676::T",
                   NEIL1 = "JH393281.1::34191182::A",
                   GABRA4 = "JH393292.1::13145599::T",
                   GRIA2 = "JH393386.1::3912673::A",
                   DKK3 = "JH393338.1::6475754::A")

dat <- read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             "A_G_alleles", "A_G_all_sites_proportions.txt.gz"))
dat <- as.data.frame(dat)
rownames(dat) <- dat[, 1]
dat <- dat[, -1]

fmat_validation <- dat[rownames(dat) %in% all_validation_sites, ]

fmat_validation <- fmat_validation * 100

# rename rows to gene names
rownames(fmat_validation) <- names(all_validation_sites[match(rownames(fmat_validation), 
                                                    all_validation_sites)])

ha2 <- HeatmapAnnotation(df = pdata_heatmap_genes[c('State', 'region')], 
                                      col = list(
                                        region = region_cols,
                                          State = state_cols
                                                 ),
                           annotation_legend_param = list(
                             State = list(title = "Sampling Group",
                                          at = state_order, 
                                          labels = state_order, nrow = 1),
                             region = list(title = "Brain Region", nrow = 3)), 
                           gap = unit(3, "mm"),
                           show_annotation_name = TRUE,
                           annotation_name_side = "left",
                           annotation_name_offset = unit(1, "mm"))

# make column order match state_order
left_join(data_frame(State = state_order), 
          pdata_heatmap_genes) %>% 
  dplyr::pull(names) -> col_names_for_hmap
ht <- Heatmap(fmat_validation,
          show_row_names = T,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = F,
          cluster_columns = F,
          column_order = col_names_for_hmap,
          top_annotation = ha2,
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left")

pdf("edits/heatmap_for_validation_sites.pdf")
draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left")
dev.off()

#fig 3b
file.copy("edits/heatmap_for_validation_sites.pdf",
          figs_dir[3], overwrite = T)
```

### subset of editing sites

```{r subset of impactful sites}

all_validation_sites <- c(
                   `ZCCHC8\nsplice_site_variant` = "JH393368.1::2076972::A",
                   `EIF3A\nArg378Gly` = "JH393296.1::8517456::T",
                   `ZC3H18\nThr377Ala` = "JH393451.1::1059495::T",
                   `ZNF483\nAsn174Asp` = "JH393369.1::1779832::T",
                   `GABRA4\nArg307Gly` = "JH393292.1::13145599::T",
                   `AMIGO2\nArg304Gly` = "JH393322.1::5158676::T",
                   `NEIL1\nLys242Arg` = "JH393281.1::34191182::A",
                   `FBXW7\nIntronic` = "JH393499.1::2282179::T",
                   `CHD9\nIntronic` = "JH393285.1::6065218::A",
                   `*DKK3\n(-)Control` = "JH393338.1::6475754::A",
                   `*GRIA2\n(+)Control` = "JH393386.1::3912673::A"
                   )

#use unfiltered matrix of proportions to be able to plot GRIA2

dat <- read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             "A_G_alleles", "A_G_all_sites_proportions.txt.gz"))
dat <- as.data.frame(dat)
rownames(dat) <- dat[, 1]
dat <- dat[, -1]


fmat_validation <- dat[rownames(dat) %in% all_validation_sites, ]
fmat_validation <- fmat_validation * 100


# rename rows to gene names
rownames(fmat_validation) <- names(all_validation_sites[match(rownames(fmat_validation), 
                                                    all_validation_sites)])

# make column order match state_order
left_join(data_frame(State = state_order), 
          pdata_heatmap_genes) %>% 
  dplyr::pull(names) -> col_names_for_hmap

fmat_validation <- fmat_validation[names(all_validation_sites),
                                   col_names_for_hmap]

left_join(data_frame(State = state_order), 
          pdata_heatmap_genes) %>% 
  dplyr::select(names, State, region) %>% 
  as.data.frame() -> pdata_heatmap_genes

ha2 <- HeatmapAnnotation(df = pdata_heatmap_genes[c('State',
                                                 'region')], 
                                      col = list(
                                        region = region_cols,
                                          State = state_cols
                                                 ),
                           annotation_legend_param = list(
                             State = list(title = "Sampling Group",
                                          at = state_order, 
                                          labels = state_order, nrow = 1),
                             region = list(title = "Brain Region", nrow = 3)), 
                           gap = unit(3, "mm"),
                           show_annotation_name = TRUE,
                           annotation_name_side = "left",
                           annotation_name_offset = unit(1, "mm"))



ht <- Heatmap(fmat_validation,
              col = c("white", "red"),
          show_row_names = T,
          show_column_names = FALSE,
          cluster_rows = F,
          show_row_dend = F,
          cluster_columns = F,
          column_order = col_names_for_hmap,
          row_order = names(all_validation_sites),
          top_annotation = ha2,
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, 
     annotation_legend_side = "top", 
     heatmap_legend_side = "left")

pdf("edits/heatmap_subset_of_impactful_sites.pdf")
draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left")
dev.off()
   
```


### Add_in_hyperedits for Snpeff summary

```{r hyperedits, eval = F}
all_validation_sites <- c(
                   `ZCCHC8\nsplice_site_variant` = "JH393368.1::2076972::A",
                   `EIF3A\nArg235Gly` = "JH393296.1::8517456::T",
                   `ZC3H18\nThr377Ala` = "JH393451.1::1059495::T",
                   `ZNF483\nAsn174Asp` = "JH393369.1::1779832::T",
                   `GABRA4\nArg307Gly` = "JH393292.1::13145599::T",
                   `AMIGO2\nArg304Gly` = "JH393322.1::5158676::T",
                   `NEIL1\nLys242Arg` = "JH393281.1::34191182::A"
                   )

fmat_validation <- fmat[rownames(fmat) %in% all_validation_sites, ]
# rename rows to gene names
rownames(fmat_validation) <- names(all_validation_sites[match(rownames(fmat_validation), 
                                                    all_validation_sites)])

ha2 <- HeatmapAnnotation(df = pdata_heatmap_genes[c('State', 'region')], 
                                      col = list(
                                        region = region_cols,
                                          State = state_cols
                                                 ),
                           annotation_legend_param = list(
                             State = list(title = "Sampling Group",
                                          at = state_order, 
                                          labels = state_order, nrow = 1),
                             region = list(title = "Brain Region", nrow = 3)), 
                           gap = unit(3, "mm"),
                           show_annotation_name = TRUE,
                           annotation_name_side = "left",
                           annotation_name_offset = unit(1, "mm"))

# make column order match state_order
left_join(data_frame(State = state_order), 
          pdata_heatmap_genes) %>% 
  dplyr::pull(names) -> col_names_for_hmap

ht <- Heatmap(fmat_validation,
              col = c("white", "red"),
          show_row_names = T,
          show_column_names = FALSE,
          cluster_rows = F,
          show_row_dend = F,
          cluster_columns = F,
          column_order = col_names_for_hmap,
          row_order = names(all_validation_sites),
          top_annotation = ha2,
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left")

pdf("edits/heatmap_subset_of_impactful_sites_2.pdf")
draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left")
dev.off()



```

## Editing patterns per cluster

```{r editpattern}

# scale and center
cluster_data <- scale(t(all_sig_sites_matrix), center = T, scale = T)
cluster_data <- as.data.frame(cluster_data)
# tidy
cluster_data$sites <- rownames(cluster_data) 
cluster_data <- as_data_frame(cluster_data) %>% 
  tbl_df() %>% 
  gather(key = "site", value = "transformed_editing_freqs", -sites) %>% 
  dplyr::rename(library = sites)

# get pdata
clustered_annotated_data <- inner_join(cluster_data, pdata, 
                                       by = c("library" = "g_names"))

# get cluster id
cid <- data_frame(site = names(e_km@cluster),
                  cluster = e_km@cluster)

clustered_annotated_data <- inner_join(clustered_annotated_data, cid, 
                                       by = "site")
# set order
clustered_annotated_data <- mutate(clustered_annotated_data,
                                   State = factor(State, levels = state_order))

# boxplot
ggplot(clustered_annotated_data, aes(State, transformed_editing_freqs)) +
  geom_boxplot(aes(fill = State)) +
  facet_grid(~cluster) +
  scale_fill_manual(values = state_cols) +
  ylab("Editing Frequency, \nmean centered and scaled")

# summarize and show as mean +/- sd
summary_dat <- group_by(clustered_annotated_data, State, cluster) %>% 
  summarize(mean_editing = mean(transformed_editing_freqs, na.rm = T),
            sd_editing = sd(transformed_editing_freqs, na.rm = T) )

ggplot(summary_dat, aes(State, mean_editing)) +
  geom_point(aes(color = as.factor(cluster))) +
  geom_smooth(method = "loess", aes(group = as.factor(cluster),
                                    colour = as.factor(cluster))) +
#  geom_errorbar(aes(ymin = mean_editing - sd_editing, 
#                    ymax = mean_editing + sd_editing), width=.1) +
  scale_colour_brewer(name = "Editing Class", 
                      palette = "Dark2",
                      labels = c(
                        paste0("1 n = ", n1),
                        paste0("2 n = ", n2)
                      )) +
  ylab("Mean Editing Frequency") +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 18, colour = state_cols),
        legend.position = c(0.25, 0.9))

ggsave("edits/average_editing_across_samples.pdf", width = 4, height = 3.5)
```

## Editing Index per cluster
```{r}

raw_dat <- read_tsv(file.path("edits", "variant_allele_counts_by_strand","A_G_alleles/A_G_all_sites_raw_counts.txt.gz"))
raw_dat <- filter(raw_dat, edit_site %in% rownames(all_sig_sites_matrix))

count_data <- raw_dat %>% 
  dplyr::select(-c(`#chrom`:ref)) %>% 
  gather(key = "library", value = "counts", -edit_site, -base)

count_data <- count_data %>% 
  left_join(cid, by = c("edit_site" = "site"))

count_data <- count_data %>% 
  left_join(pdata, by = c("library" = "g_names"))

# set order
count_data <- mutate(count_data,
                     State = factor(State, levels = state_order))

count_data <- count_data %>% 
  spread(base, counts)

index_data <- group_by(count_data, State, cluster, region) %>% 
  summarize(editing_index = sum(G) / (sum(A) + sum(G))) %>% 
  ungroup() %>% 
  mutate(cluster = paste0("Group ", as.character(cluster)),
         cluster = ifelse(str_detect(cluster, "1"),
                          paste0(cluster, " (", formatC(n1, big.mark = ","), ")"),
                          paste0(cluster, " (", formatC(n2, big.mark = ","), ")")))
  

ggplot(index_data, aes(State, editing_index)) +
  geom_point(aes(color = region), show.legend = T) +
  geom_smooth(method = "loess", aes(group = as.factor(cluster)), 
              show.legend = F, 
              color = "black") +
  facet_wrap(~cluster) + 
  scale_colour_manual(values = region_cols) +
  ylab("Editing Index") +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 12, colour = state_cols),
        legend.position = "top",
        legend.title = element_blank())

ggsave("edits/editing_index_across_states_by_cluster.pdf", width = 4, height = 3.5)

#fig 2c
file.copy("edits/editing_index_across_states_by_cluster.pdf", figs_dir[2], overwrite = T)

index_data <- group_by(count_data, State, region) %>% 
  summarize(editing_index = sum(G) / (sum(A) + sum(G))) %>% 
  ungroup() 
  

ggplot(index_data, aes(State, editing_index, group = region)) +
  geom_point(aes(colour= region)) +
  geom_smooth(method = "loess", 
              aes(color = region)) +
  scale_colour_brewer(palette = "Dark2") +
  ylab("Editing Index") +
  expand_limits(y = 0) + 
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 12, colour = state_cols),
        legend.position = "top",
        legend.title = element_blank(),
        legend.key = element_rect(colour = "white"))

ggsave("edits/editing_index_across_states_all_clusters.pdf", width = 4, height = 3.5)
file.copy("edits/editing_index_across_states_all_clusters.pdf", figs_dir[2], overwrite = T)
```


## pvalue distribution per class

```{r}
fdr_all_sites %>% 
  inner_join(cid, by = "site") -> cluster_fdrs


cluster_fdrs %>% 
  dplyr::filter(., 
                site %in% sites_no_excessive_na) -> cluster_fdrs

ggplot(cluster_fdrs, aes(FDR)) +
  geom_density(aes(fill = as.factor(cluster)), alpha = 0.5)

cluster_fdrs <- dplyr::rename(cluster_fdrs, 
                              kmeans_cluster = cluster)

# coerce numeric bed coords to integer to avoid scientific notation printing causing 
# issues with downstream tools
# see https://github.com/tidyverse/readr/issues/671

cluster_fdrs <- mutate_at(cluster_fdrs, c("start",
                                          "end"), as.integer)

write_tsv(cluster_fdrs, "edits/A_G_filtered_fdr0.01_sites_annotation_kmeans.txt")
R.utils::gzip( "edits/A_G_filtered_fdr0.01_sites_annotation_kmeans.txt", remove = T, overwrite = T)

# write as strict bed file also
cluster_fdrs_bed <- cluster_fdrs %>% 
  dplyr::select(chrom, 
                start, 
                end,
                kmeans_cluster,
                alt,
                strand) %>% 
  unique()

write_tsv(cluster_fdrs_bed, 
          "edits/A_G_filtered_fdr0.01_sites.bed",
          col_names = F)

```

 
## Comparing editing sites to counts

```{r}
count_dat <- raw_dat
count_dat$total <- rowSums(count_dat[, c(6:ncol(count_dat))])
count_dat_summary <- count_dat %>% 
  dplyr::select(edit_site, total) %>% 
  group_by(edit_site) %>% 
  summarize(total_coverage = sum(total))

count_dat_summary_sigedits <- count_dat_summary %>% 
  filter(edit_site %in% km_sites_1) 

class1 <- fdr_all_sites %>% 
  filter(site %in% km_sites_1)

count_dat_summary_sigedits <- inner_join(count_dat_summary_sigedits, 
                                         class1, by = c("edit_site" = "site")) %>% 
  dplyr::select(edit_site, total_coverage, FDR) %>% 
  group_by(edit_site, total_coverage) %>% 
  summarize(FDR = min(FDR)) %>% 
  mutate(FDR = -log10(FDR)) %>% 
  ungroup()


ggplot(count_dat_summary_sigedits, aes(FDR, log10(total_coverage))) +
  geom_point() +
  geom_smooth(method = "lm")

class1_edit_freq <- dat[rownames(dat) %in% km_sites_1, ] 
mean_edit_freq <- rowMeans(class1_edit_freq, na.rm = T)

```



## Recompute genomic distribution for sig versus not-sig sites

Incorporate both ensembl predictions (for CDS) and transcriptome build

```{r get_genomic_distribution_of_sites}
annotation <- read_tsv(file.path("edits", 
                                 "variant_allele_counts_by_strand", 
                                 "A_G_alleles", "A_G_filtered_sites_annotation.txt.gz"))
annotated_sig <- annotation %>% filter(FDR < 0.01)


dplyr::select(annotated_sig, site, EFF) %>%
  unique() %>% 
  mutate(., 
       EFF = str_split(EFF, ",") %>%
         map_chr(.,
                 ~unique(.x) %>%
                   .[. != "custom"] %>% 
                   paste0(., collapse = ","))) %>% 
  group_by(EFF) %>% 
  summarize(total = n()) %>%
  dplyr::arrange(desc(total))

```


```{r valr, eval = F}
sig_bed <- dplyr::select(annotated_sig, chrom, start, end, site)
bed <- read_tsv(file.path(db_dir, "ensembl85", "Ictidomys_tridecemlineatus.spetri2.85.bed.gz"),
                col_names = F)
colnames(bed) <- c("chrom", "start", "end", "name", "score", "strand", 
                   "geneid", "biotype", "transcript", "genename",
                   "exon_count", "feature")

tmp_bed <- bed_intersect(sig_bed, bed)
not_bed <- bed_intersect(sig_bed, bed, invert = T)

features_bed <- bind_rows(tmp_bed, not_bed)

#replace site.x col with site col after binding up non-intersecting sites
features_bed <- mutate(features_bed, site.x = ifelse(!is.na(site),
                                            site,
                                            site.x))
features_bed <- features_bed %>% 
  group_by(site.x) %>% 
  summarize(features = paste0(unique(feature.y), collapse = ","), 
            genes = paste0(unique(geneid.y), collapse = ","))

best_bed <- mutate(features_bed, 
                   best_features = ifelse(str_detect(features, "CDS"), "CDS",
                                     ifelse(str_detect(features, "three_prime_utr"), "3'UTR",
                                            ifelse(str_detect(features, "five_prime_utr"), "5'UTR", 
                                                   ifelse(str_detect(features, "start_codon"), "CDS", 
                                                          ifelse(str_detect(features, "stop_codon"), "CDS", 
                                                                 ifelse(str_detect(features, "exon"), "ncRNA",
                                                                 ifelse(str_detect(features, "gene"), "Intron", "Intergenic")))))))) 
                                                                                                               
summary_bed <- best_bed %>% group_by(best_features) %>%  summarize(totals = n())

p <- ggplot(summary_bed, aes(best_features, totals)) +
  geom_bar(stat = "identity") +
  ylab("Number of Significant Sites") +
  theme(
    axis.title.x = element_blank()
  )
p
save_plot("edits/genomic_distribution.pdf", p, base_width = 5)

bed <- read_tsv(file.path(db_dir, "ensembl85", "Ictidomys_tridecemlineatus.spetri2.85.bed.gz"),
                col_names = F)

colnames(bed) <- c("chrom", "start", "end", "name", "score", "strand", "geneid",
  "biotype", "transcript", "genename", "exon_count", "feature")
exon_bed <- filter(bed, feature == "exon")
transcript_bed <- filter(bed, feature == "transcript")
cds_bed <- filter(bed, feature == "CDS" | feature == "start_coding" | feature == "stop_codon")
three_bed <- filter(bed, feature == "three_prime_utr")
five_bed <- filter(bed, feature == "five_prime_utr")

ncRNA <- anti_join(exon_bed, cds_bed, by = "transcript") %>%  
  anti_join(., three_bed, by = "transcript") %>%  
  anti_join(., five_bed, by = "transcript")

transcript_bed <- group_by(transcript_bed, strand)
exon_bed <- group_by(exon_bed, strand)
introns <- bed_subtract(transcript_bed, exon_bed)
regions <- list(introns, cds_bed, three_bed, five_bed, ncRNA)
regions <- map(regions, ~mutate(.x, length = end - start))
regions <- map(regions, ~ungroup(.x) %>% summarize(total = sum(length)))
names(regions) <- c("intron", "CDS", "3'UTR", "5'UTR", "ncRNA")
regions <- bind_rows(regions, .id = "region")

p <- ggplot(regions, aes(region, total)) +
  geom_bar(stat = "identity") +
  ylab("Nucleotides found in each Annotation") +
  theme(
    axis.title.x = element_blank()
  )
p
save_plot("edits/annotation_genomic_distribution.pdf", p, base_width = 5)
```

### repeat distribution

```{r, eval = F}
dl_file <- file.path(db_dir, "ensembl85", "ensembl85_rmsk.txt.gz")
if(!file.exists(dl_file)){
  url <- "http://hgdownload.soe.ucsc.edu/goldenPath/speTri2/database/rmsk.txt.gz"
  download.file(url, dl_file)
}

bed_file <- file.path(db_dir, "ensembl85", "ensembl85_rmsk.bed.gz")

if(!file.exists(bed_file)){
  dat <- read_tsv(dl_file, col_names = F)
  dat <- dplyr::select(dat, 
                       X6, #chrom 
                       X7, #start
                       X8, #end
                       X11, #repeat name (e.g. STRIDE)
                       X12, # repeat class (e.g. SINE)
                       X10 #strand
                       ) %>% 
    dplyr::rename(chrom = X6,
                  start = X7,
                  end = X8,
                  name = X11,
                  class = X12,
                  strand = X10)
  write_tsv(dat, bed_file)
}
```
```{r analyze_repeats, eval = F}
bed <- read_tsv(file.path(db_dir, "ensembl85", "ensembl85_rmsk.bed.gz"))

new_chroms <- str_replace(sig_bed$chrom, ".1$", "")
sig_bed_ucsc <- sig_bed
sig_bed_ucsc$chrom <- new_chroms
rpt_bed <- bed_intersect(sig_bed_ucsc, bed)
non_rpt_bed <- bed_intersect(sig_bed_ucsc, bed, invert = T)

rpt_bed <- rpt_bed %>% 
  dplyr::select(site.x, name.y, class.y) %>% 
  mutate(repeat_found = "repeat")

non_rpt_bed <- non_rpt_bed %>%  dplyr::select(site) %>% 
  mutate(repeat_found = "non-repeat", name.y = "none", class.y = "none", site.x = site) %>%
  dplyr::select(-site, site.x, name.y, class.y, repeat_found)

rpt_bed <- bind_rows(rpt_bed, non_rpt_bed)

rpt_bed <- rpt_bed %>% 
  group_by(site.x) %>%  
  summarize(repeat_names = valr::values_unique(name.y), 
            repeat_classes = valr::values_unique(class.y),
            repeat_present = first(repeat_found))

#annotate with best_bed from above
rpt_bed <- left_join(rpt_bed, best_bed)                                                                                              
# add in a total factor
tmp <- rpt_bed %>% mutate(best_features = "Total")
rpt_bed <- bind_rows(rpt_bed, tmp)

# add in totals for text
rpt_txt <- rpt_bed %>% 
  group_by(best_features) %>% 
  summarize(total_n = paste0("n = ", 
                             formatC(n(), big.mark = ",")))

rpt_bed <- mutate(rpt_bed, 
                  best_features = factor(best_features, 
                                                  levels = c("Total",
                                                             "Intron",
                                                             "Intergenic",
                                                             "CDS",
                                                             "5'UTR",
                                                             "3'UTR",
                                                             "ncRNA")),
                  repeat_present = factor(repeat_present,
                                          levels = c("repeat", "non-repeat")))

p <- ggplot(rpt_bed, aes(best_features)) +
  geom_bar(aes(fill = repeat_present), 
           position = "fill",
           color="black") +
  geom_text(data = rpt_txt, 
            aes(best_features, 0.8, label = total_n),
            angle = 90,
            size = 5) +
  ylab("Proportion of editing sites") +
  scale_fill_manual(values = c("#D9D9D9", "#525252")) +
  theme(
    axis.title.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top"
  )
p

save_plot("edits/repeat_distribution.pdf", p, base_width = 6)

#fig 3
file.copy("edits/repeat_distribution.pdf", figs_dir[3], overwrite = T)

sum_bed <- mutate(rpt_bed, 
                  repeats_single_assignment = ifelse(str_count(repeat_classes, ",")  > 0, "multiple_annotations", repeat_classes)) %>% 
  dplyr::filter(best_features != "Total")

ggplot(sum_bed, aes(repeat_present)) +
  geom_bar(aes(fill = repeat_present)) +
  ylab("variants called (n)") +
  scale_fill_manual(values = color_fxn(12)) +
  theme(
    axis.title.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "none"
  )
p
```



## Heatmap of conserved (found in mouse or human) sites

```{r, conserved_sites, fig.width = 6, fig.height = 8}

mouse_sig <- read_tsv("edits/shared_mouse_sig_edits.txt")
human_sig <- read_tsv("edits/shared_human_sig_edits.txt")

# might be useful to write these out to a table also with FDR values

mouse_sig %>% dplyr::select(gene.x, site.y, annot2.x) -> sig_conserved
bind_rows(sig_conserved, 
          human_sig %>% 
            dplyr::select(gene.x, site.y, annot2.x)) -> sig_conserved

dplyr::mutate(sig_conserved, gene.x = str_to_upper(gene.x)) %>% 
  unique() %>% 
  arrange(gene.x) %>% 
  mutate(gene_effect_ids = paste0(gene.x, "::", annot2.x)) -> sig_conserved

# make unique names for each site
sig_conserved %>% 
  group_by(gene_effect_ids) %>% 
  mutate(n = n(), 
         row_id = row_number(), 
         gene_effect_ids_unique = ifelse(n > 1, 
                                         paste0(gene_effect_ids, "::site", row_id), 
                                         gene_effect_ids)) -> sig_conserved

sig_conserved_sites <- dplyr::pull(sig_conserved, site.y)
names(sig_conserved_sites) <- dplyr::pull(sig_conserved, gene_effect_ids_unique)

fmat <- format_heatmap_matrix(dat)
fmat_validation <- fmat[rownames(fmat) %in% sig_conserved_sites, ]

# rename rows to gene names with potential impact of site
row_geneids <- names(sig_conserved_sites[match(rownames(fmat_validation), 
                                                    sig_conserved_sites)])
rownames(fmat_validation) <- row_geneids

ha2 <- HeatmapAnnotation(df = as.data.frame(pdata[c('State', 'region')]), 
                                      col = list(
                                        "region" = region_cols,
                                          "State" = state_cols
                                                 ),
                           annotation_legend_param = list(
                             State = list(title = "Sampling Group",
                                          at = state_order, 
                                          labels = state_order, nrow = 1),
                             region = list(title = "Brain Region", nrow = 3)), 
                             gap = unit(3, "mm"),
                            show_annotation_name = TRUE,
                            annotation_name_side = "left",
                            annotation_name_offset = unit(1, "mm"))


ht <- Heatmap(fmat_validation,
          show_row_names = T,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = F,
          cluster_columns = T,
         # column_order = col_names_for_hmap,
          top_annotation = ha2,
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left")

pdf("edits/heatmap_for_conserved_sites.pdf")
draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left")
dev.off()

#fig 3b
file.copy("edits/heatmap_for_conserved_sites.pdf",
          figs_dir[2], overwrite = T)
```
  



## Heatmap of non-A-to-G significant variants

```{r nonsig}

# editing frequencies
dat <- read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             "A_G_alleles", "A_G_filtered_sites_proportions.txt.gz"))
dat <- as.data.frame(dat)
rownames(dat) <- dat[, 1]
dat <- dat[, -1]

# convert to percentages
dat <- dat * 100


non_ag <- list(
              "A_C_alleles",
              "A_T_alleles",
              "C_A_alleles",
              "C_G_alleles",
              "C_T_alleles",
              "G_A_alleles",
              "G_C_alleles",
              "G_T_alleles",
              "T_A_alleles",
              "T_C_alleles",
              "T_G_alleles")

ref_alt <- map_chr(non_ag, 
                   ~str_split(.x, "", simplify = T) %>%
                     .[1:3] %>% 
                     paste0(., collapse = ""))

# editing frequencies
dat <- map2(non_ag, ref_alt,
           ~read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             .x, paste0(.y, "_filtered_sites_proportions.txt.gz"))))
convert_to_percent <- function(df){
  df <- as.data.frame(df)
  rownames(df) <- df[, 1]
  df <- df[, -1]

  # convert to percentages
  df <- df * 100
  df
}

dat <- map(dat, convert_to_percent)

annots <- map2(non_ag, ref_alt, 
               ~read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             .x, paste0(.y, "_filtered_sites_annotation.txt.gz"))))

fdr <- map(annots, ~filter(.x, FDR < 0.01))

# EFF column contains predictions based on ensembl85
fdr <- map(fdr, ~dplyr::rename(.x, EFF_ensembl = EFF))

sig_sites <- map(fdr, ~dplyr::pull(.x, site) %>% unique())
all_sig_sites_matrix <- map2(dat, sig_sites, ~.x[rownames(.x) %in% .y, ] %>% as.matrix)

names(all_sig_sites_matrix) <- str_replace(ref_alt, "_", "->") %>% 
  str_c(., " variants")

pdata_file = file.path(docs_dir, "BrainRegionRNAseqLibMetadata.txt")
 # make_groups
g_names <- map(dat, ~colnames(.x)) %>% unlist()
animal_number <- str_extract(g_names, "[MHB][0-9]+")
  
 # allele_type <- rep(c("A", "G"), each = 30)
groups <- data_frame(g_names, animal_number)
  
pdata <- readr::read_tsv(pdata_file)
  
pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)

pdata <- mutate(pdata,
                region = ifelse(region == "Forebrain",
                                "Cerebrum",
                                region))

pdata <- mutate(pdata, 
                  abbrev_sample = str_split(sample, "_", simplify = T)[, 1])
  
pdata <- inner_join(groups, 
                      pdata, 
                      by = c("animal_number" = "abbrev_sample")) %>%
    dplyr::select(g_names, animal_number, State, region)

plot_nonag <- function(mat, pdata, c_title){
  pdata_heatmap <- data_frame( "names" = colnames(mat)) %>%
        inner_join(., pdata, by = c("names" = "g_names")) %>%
        dplyr::select(names, State, region) %>% as.data.frame() %>% 
        arrange(State)
  
  reordered_cols <- pdata_heatmap$names %>% unlist()
  
  mat <- mat[, reordered_cols]
  
  mat <- format_heatmap_matrix(mat)
  nsites <- nrow(mat)
  
  ha2 <- HeatmapAnnotation(df = pdata_heatmap[, c('State', 'region')], 
                                        col = list(
                                          region = region_cols,
                                            State = state_cols
                                                   ),
                             annotation_legend_param = list(
                               State = list(title = "Hiberation\nState",
                                            at = state_order, labels = state_order)))
  
  
  ht <- Heatmap(mat,
        #        col = circlize::colorRamp2(c(0, 1), c("white", "red")),
            show_row_names = FALSE,
            show_column_names = FALSE,
            cluster_rows = T,
            cluster_columns = T,
            show_row_dend = T,
            top_annotation = ha2,
            row_title = "Editing Sites",
            column_title = paste(c_title, "n =", nsites), 
            heatmap_legend_param = list(
              title = "Percent\nEditing (%)")
           )
  
  draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")
}
walk2(all_sig_sites_matrix, names(all_sig_sites_matrix), 
      ~plot_nonag(.x, pdata, .y))

walk2(all_sig_sites_matrix, names(all_sig_sites_matrix), 
      function(x, y) {
        pdf(paste0("edits/", y, "_heatmap.pdf"))
        plot_nonag(x, pdata, y)
        dev.off()
      }
)

map(all_sig_sites_matrix, 
    ~format_heatmap_matrix(.x) %>% nrow(.)) %>% 
  unlist() %>% 
  sum() -> non_A_to_G_sites
```


```{r, eval = F}
annots <- read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             "A_G_alleles", "A_G_filtered_sites_annotation.txt.gz"))

high_fdr <- filter(annots, FDR >= 0.01)

# get list of non-NA sites 




map(all_sig_sites_matrix, 
    ~format_heatmap_matrix(.x) %>% rownames(.)) %>% 
  unlist() %>% 
  unname() -> non_A_to_G_sites_ids

write_tsv(high_fdr, "edits/A_G_filtered_notsignificant_sites_annotation_kmeans.txt")
R.utils::gzip("edits/A_G_filtered_notsignificant_sites_annotation_kmeans.txt", 
              remove = T, overwrite = T)

# write as strict bed file also
high_fdrs_bed <- high_fdr %>% 
  dplyr::select(chrom, 
                start, 
                end,
                REF,
                ALT,
                strand) %>% 
  unique()

write_tsv(high_fdrs_bed, 
          "edits/A_G_filtered_high_fdr_sites.bed",
          col_names = F)
```

There are only `r non_A_to_G_sites` non-A-to-G sites. 