---
title: "make_geo_data"
author: "Kent Riemondy RBI"
date: "11/14/2017"
output: html_document
---

```{r setup, include=FALSE}
source("globals.R")
```

## Format processed data for GEO

### Annotation information for all EdgeR tested editing sites
```{r}
dat <- read_tsv("edits/variant_allele_counts_by_strand/A_G_alleles/A_G_filtered_sites_annotation.txt.gz")

out_dat <- dplyr::select(dat,
                         Region:Disruption,
                         GENE_ID:GENE_NAME)

#simplify SNPeff annotations
out_dat %>% 
  mutate(EFF = str_replace(EFF, ",custom", "") %>% 
           str_split(., ",") %>% 
           map(., unique) %>%
           map_chr(., ~paste0(.x, collapse = ",")),
         Disruption = str_split(Disruption, ",") %>% 
           map(., unique) %>%
           map_chr(., ~paste0(.x, collapse = ","))) -> out_dat

dir.create("edits/geo/", recursive = T)
#write out as gzipped
write_gztsv(out_dat, "edits/geo/gatk_sites_annotation.txt")

cnames <- colnames(out_dat)
descript <- c(
  "brain region",
  "site id (contig::position(1-based)::ref_nucleotide)",
  "LogFC of alternate allele in Ar versus LT",
  "LogFC of alternate allele in SpD versus LT",
  "LogFC of alternate allele in IBA versus LT",
  "LogFC of alternate allele in Ent versus LT",
  "LogFC of alternate allele in SA versus LT",
  "logCPM (average across all states)",
  "LR (test statistic)",
  "PValue from edgeR LRT test",
  "FDR (Benjamini-Hochberg)",
  "contig",
  "start 0-based",
  "end 0-based",
  "REF nucleotide",
  "ALT nucleotide",
  "strand",
  "EFF predicted effect from snpEFF",
  "predicted distruption from snpEFF",
  "Gene id from transcriptome buid",
  "biotype from Ensembl annotations",
  "Ensembl transcript Id",
  "Gene Name")

gatk_annotation_descript <- data_frame(cnames,
                                       descript)
  
```

### Editing proportions for EdgeR tested editing sites

```{r}
dat <- read_tsv("edits/variant_allele_counts_by_strand/A_G_alleles/A_G_filtered_sites_proportions.txt.gz")

file.copy("edits/variant_allele_counts_by_strand/A_G_alleles/A_G_filtered_sites_proportions.txt.gz", 
          "edits/geo/gatk_sites_proportions.txt.gz", 
          overwrite = T)

cnames <- c("edit_site",
            "individual libraries")
descript <- c(
  "site id (contig::position(1-based)::ref_nucleotide)",
  "Editing proportions (G / G + A) for all 90 libraries, missing data is NaN")

gatk_proportion_descript <- data_frame(cnames,
                                       descript)
  
```


### read count for EdgeR tested editing sites

```{r}
dat <- read_tsv("edits/variant_allele_counts_by_strand/A_G_alleles/A_G_filtered_sites_raw_counts.txt.gz")

out_dat <- dplyr::select(dat,
                         edit_site, 
                         base:B84_7_AGTCAA_L001_counts.txt.gz) %>% 
  dplyr::rename(nucleotide = base)
cnames <- c("edit_site",
            "nucleotide",
            "individual libraries")

descript <- c(
  "site id (contig::position(1-based)::ref_nucleotide)",
  "counted allele (A = reference, G = edited)",
  "Read counts for all 90 libraries, missing data is 0")

gatk_counts_descript <- data_frame(cnames,
                                       descript)
  
#write out as gzipped
write_gztsv(out_dat, "edits/geo/gatk_sites_raw_counts.txt")
```

## Hyperedited sites

### bedfile
```{r}
hyper_edits <- read_tsv("hyperedits/all_hyperedits.bed.gz")
hyper_edits %>% 
  dplyr::select(chrom, start, end, site_id, supporting_reads, strand) %>%
  unique() -> out_dat

cnames <- colnames(out_dat)
descript <- c(
  "contig",
  "start (0-based)",
  "end (0-based)",
  "site id (contig::position(1-based)::ref_nucleotide)",
  "number of reads supporting the hyperediting site",
  "strand")

hyper_bed_descript <- data_frame(cnames,
                                       descript)
  
write_gztsv(out_dat,
          "edits/geo/hyperedited_sites.bed")
```


### Annotation information for hyperedited sites
```{r}
dat <- read_tsv(file.path("hyperedits", "diffedits",
                             "A_G_filtered_sites_annotation.txt.gz"))

out_dat <- dplyr::select(dat,
                         Region:Disruption,
                         GENE_ID:GENE_NAME)

#simplify SNPeff annotations
out_dat %>% 
  mutate(EFF = str_replace(EFF, ",custom", "") %>% 
           str_split(., ",") %>% 
           map(., unique) %>%
           map_chr(., ~paste0(.x, collapse = ",")),
         Disruption = str_split(Disruption, ",") %>% 
           map(., unique) %>%
           map_chr(., ~paste0(.x, collapse = ","))) -> out_dat


#write out as gzipped
write_gztsv(out_dat, "edits/geo/hyperedited_sites_annotation.txt")

cnames <- colnames(out_dat)
descript <- c(
  "brain region",
  "site id (contig::position(1-based)::ref_nucleotide)",
  "LogFC of alternate allele in Ar versus LT",
  "LogFC of alternate allele in SpD versus LT",
  "LogFC of alternate allele in IBA versus LT",
  "LogFC of alternate allele in Ent versus LT",
  "LogFC of alternate allele in SA versus LT",
  "logCPM (average across all states)",
  "LR (test statistic)",
  "PValue from edgeR LRT test",
  "FDR (Benjamini-Hochberg)",
  "contig",
  "start 0-based",
  "end 0-based",
  "REF nucleotide",
  "ALT nucleotide",
  "strand",
  "EFF predicted effect from snpEFF",
  "predicted distruption from snpEFF",
  "Gene id from transcriptome buid",
  "biotype from Ensembl annotations",
  "Ensembl transcript Id",
  "Gene Name")

hyper_annotation_descript <- data_frame(cnames,
                                       descript)
  
```

### Editing proportions for hyper tested editing sites

```{r}
file.copy("hyperedits/diffedits/A_G_filtered_sites_proportions.txt.gz", 
          "edits/geo/hyperedited_sites_proportions.txt.gz", 
          overwrite = T)

cnames <- c("edit_site",
            "individual libraries")
descript <- c(
  "site id (contig::position(1-based)::ref_nucleotide)",
  "Editing proportions (G / G + A) for all 90 libraries, missing data is NaN")

hyper_proportion_descript <- data_frame(cnames,
                                       descript)
  
```

### read count for hyper tested editing sites

```{r}
dat <- read_tsv("hyperedits/diffedits/A_G_filtered_sites_raw_counts.txt.gz")

out_dat <- dplyr::select(dat,
                         edit_site, 
                         base:B84_7_AGTCAA_L001_counts.txt.gz) %>% 
  dplyr::rename(nucleotide = base)
cnames <- c("edit_site",
            "nucleotide",
            "individual libraries")

descript <- c(
  "site id (contig::position(1-based)::ref_nucleotide)",
  "counted allele (A = reference, G = edited)",
  "Read counts for all 90 libraries, missing data is 0")

hyper_counts_descript <- data_frame(cnames,
                                       descript)
  
#write out as gzipped
write_gztsv(out_dat, "edits/geo/hyperedited_sites_raw_counts.txt")
```

## Transcriptome

```{r}
R.utils::gzip(file.path(db_dir, "denovo_annotation", "fixed.all95.gtf"), 
              remove = FALSE,
              overwrite = TRUE)

file.copy(file.path(db_dir, "denovo_annotation", "fixed.all95.gtf.gz"),
          "edits/geo/fixed.all95.gtf.gz")
tx_descript <- c("transcriptome in gtf form built with stringtie and merged with TACO")
```

## README

```{r}

files <- dir("edits/geo")

out_ids <- list( 
  data_frame(cname = NA,
             descript = tx_descript),
  gatk_annotation_descript,
  gatk_proportion_descript,
  gatk_counts_descript,
  hyper_annotation_descript,
  hyper_proportion_descript,
  hyper_counts_descript,
  hyper_bed_descript
  )

names(out_ids) <- files

file_descript <- c(
  "transcriptome (GTF)",
  "RNA editing sites identified by GATK approach and tested by EdgeR for hibernation specific editing (TXT)",
  "editing proportions for RNA editing sites identified by GATK approach (TXT)",
  "edited and non-edited read counts for RNA editing sites identified by GATK approach (TXT)",
   "RNA editing sites identified by hyper-editing approach and tested by EdgeR for hibernation specific editing (TXT)",
  "editing proportions for RNA editing sites identified by hyper-editing approach (TXT)",
  "edited and non-edited read counts for RNA editing sites identified by hyper-editing approach (TXT)",
  "BED file of sites identified by hyperediting approach (BED)"
)

file_descriptions <- data_frame(filename = files, 
                                file_description = file_descript)


out_ids %>% 
  bind_rows(., .id = "filename") -> out

out %>% 
  left_join(file_descriptions,
            by = "filename") %>% 
  dplyr::rename(column_names = cnames,
                description = descript) %>% 
  dplyr::select(filename, file_description, column_names, description) -> out
write_tsv(out, "edits/geo/readme.txt")
```

## Supplemental Table 2

Combine annotation data with kmeans information and generate a significant and constitutive site table.

```{r}

gatk_sites <- read_tsv("edits/A_G_filtered_fdr0.01_sites_annotation_kmeans.txt.gz")
gatk_sites <- dplyr::filter(gatk_sites, 
                            kmeans_cluster == 1) %>% 
  dplyr::select(-EFF,
                Region:Disruption,
                Nucleotide:AminoAcid,
                GENE_ID:GENE_NAME, 
                kmeans_cluster) %>% 
    dplyr::rename(EFF = EFF_ensembl) %>% 
  unique()

hyperedited_sites <- read_tsv("hyperedits/diffedits/A_G_filtered_sites_annotation_kmeans.txt.gz")
hyperedited_sites <- hyperedited_sites %>% 
  dplyr::filter(kmeans_cluster != 3, 
                FDR < 0.01) %>% 
    dplyr::select(Region:Disruption,
                Nucleotide:AminoAcid,
                GENE_ID:GENE_NAME, 
                kmeans_cluster) %>% 
  unique()

both_sites <- list(hyperedited_sites,
                   gatk_sites)
names(both_sites) <- c("hyperedited", "gatk")

# change kmeans cluster in hyper from 1,2 to 3,4
both_sites$hyperedited <- mutate(both_sites$hyperedited,
                                 kmeans_cluster = ifelse(kmeans_cluster == 1, 
                                                         3,
                                                         ifelse(kmeans_cluster == 2,
                                                                4,
                                                                NA)))
# annotate ones found in both gatk and hyper
intersect(both_sites$hyperedited$site,
          both_sites$gatk$site) %>% unique() -> shared_sig_sites

both_sites$hyperedited <- both_sites$hyperedited[!both_sites$hyperedited$site %in% shared_sig_sites, ]

cols <- colnames(both_sites$hyperedited)
both_sites$gatk <- both_sites$gatk %>% 
  dplyr::select(one_of(cols))


both_sites <- bind_rows(both_sites, .id = "edit_type")
both_sites <- mutate(both_sites, 
                     edit_type = ifelse(site %in% shared_sig_sites,
                                        "both",
                                        edit_type))

hyper_edits <- read_tsv("hyperedits/diffedits/A_G_filtered_sites_annotation_kmeans.txt.gz")
hyper_edits_non_sig <- hyper_edits %>% 
  dplyr::filter(FDR > 0.5) %>% 
  dplyr::select(Region:Disruption,
                Nucleotide:AminoAcid,
                GENE_ID:GENE_NAME) %>% 
  unique()

hyper_edits_non_sig <- hyper_edits_non_sig[!hyper_edits_non_sig$site %in% both_sites$site, ] 

gatk_non_sigsites <- read_tsv("edits/A_G_filtered_notsignificant_sites_annotation_kmeans.txt.gz")
gatk_non_sigsites <- gatk_non_sigsites %>%
  dplyr::filter(FDR > 0.5) %>% 
  dplyr::select(Region:Disruption,
                Nucleotide:AminoAcid,
                GENE_ID:GENE_NAME) %>% 
  unique()

gatk_non_sigsites <- gatk_non_sigsites[!gatk_non_sigsites$site %in% both_sites$site, ] 

semi_join(gatk_non_sigsites, 
          hyper_edits_non_sig, 
          by = "site") -> both_nonsig


both <- list(both_sites, both_nonsig)
names(both) <- c("Cold-Enriched Sites", "Constitutive Sites")

#simplify SNPeff annotations
map(both,
    ~.x %>% 
  mutate(EFF = str_replace(EFF, "(custom|,custom)", "") %>% 
           str_split(., ",") %>% 
           map(., unique) %>%
           map_chr(., ~paste0(.x, collapse = ",")),
         Disruption = str_split(Disruption, ",") %>% 
           map(., unique) %>%
           map_chr(., ~paste0(.x, collapse = ",")),
         AminoAcid = str_split(AminoAcid, ",") %>% 
           map(., unique) %>%
           map_chr(., ~paste0(.x, collapse = ",")),
         Nucleotide = str_split(Nucleotide, ",") %>% 
           map(., unique) %>%
           map_chr(., ~paste0(.x, collapse = ",")))) -> out_dat

cnames <- colnames(out_dat$`Cold-Enriched Sites`)
descript <- c(
  "RNA editing detection approach (cold-enriched sites only)",
  "brain region",
  "site id (contig::position(1-based)::ref_nucleotide)",
  "LogFC of alternate allele in Ar versus LT",
  "LogFC of alternate allele in SpD versus LT",
  "LogFC of alternate allele in IBA versus LT",
  "LogFC of alternate allele in Ent versus LT",
  "LogFC of alternate allele in SA versus LT",
  "logCPM (average across all states)",
  "LR (test statistic)",
  "PValue from edgeR LRT test",
  "FDR (Benjamini-Hochberg)",
  "contig",
  "start 0-based",
  "end 0-based",
  "REF nucleotide",
  "ALT nucleotide",
  "strand",
  "EFF predicted effect from snpEFF",
  "predicted distruption from snpEFF",
  "Nucleotide change predicted by snpEFF",
  "AminoAcid change predicted by snpEFF",
  "Gene id from transcriptome buid",
  "biotype from Ensembl annotations",
  "Ensembl transcript Id",
  "Gene Name",
  "kmeans_cluster (cold-enriched sites only)")

out_descript <- data_frame(Columns = cnames,
                                        Description = descript)

map(out_dat, ~set_xlsx_class(.x,
               c("GENE_ID", "TRANSCRIPT_ID", "GENE_NAME"),
               "Text")) -> out_dat

map(out_dat, ~set_xlsx_class(.x, 
               c("start", "end", "GENE_NAME"),
               "Number")) -> out_dat

map(out_dat, ~set_xlsx_class(.x, 
               c(str_subset(colnames(out_dat$`Cold-Enriched Sites`), "log"),
                 "LR", "PValue", "FDR"),
               "Scientific")) -> out_dat

out_dat <- list(out_descript,
                out_dat$`Cold-Enriched Sites`,
                out_dat$`Constitutive Sites`)
names(out_dat) <- c("Description", "Cold-Enriched Sites", "Constitutive Sites")

openxlsx::write.xlsx(out_dat, "Supplemental-Table-2.xlsx")
```


There are `r out_dat[[1]] %>% pull(GENE_NAME) %>% unique() %>% na.omit(.)` unique genes in the torpor-enriched sites. 