---
title: "compare spring warm versus winter warm"
author: "Kent Riemondy RBI"
date: "12/19/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```

## Test for editing differences between warm in spring/summer and winter
```{r}
source("globals.R")
library(scales)
```

```{r get_pdata}

get_groups <- function(matrix_dat, regex_for_fastq, 
                       ref_nt, alt_nt, 
                       pdata_file = file.path(docs_dir,
                                              "BrainRegionRNAseqLibMetadata.txt")) {
 # make_groups
  g_names <- colnames(matrix_dat)
  animal_number <- str_extract(g_names, regex_for_fastq)
  
  allele_type <- rep(c(ref_nt, alt_nt), each = 30)
  groups <- data_frame(g_names, animal_number, allele_type)
  
  pdata <- readr::read_tsv(pdata_file)
  
  pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)
  
  pdata <- mutate(pdata, 
                  abbrev_sample = str_split(sample, "_", simplify = T)[, 1])
  
  pdata <- inner_join(groups, 
                      pdata, 
                      by = c("animal_number" = "abbrev_sample")) %>%
    dplyr::select(g_names, animal_number, allele_type, State)
  
  pdata <- mutate(pdata,
                  temp = ifelse(str_detect(State,
                                           "IBA|Ent"),
                                "winter_warm",
                                ifelse(str_detect(State,
                                           "SA|SpD"),
                                "summer_warm", "cold")))
    
  pdata
}

regexes <- paste0(c("M", "H", "B"), "[0-9]+")

mat <- read_tsv(file.path("edits",
                          "variant_allele_counts_by_strand",
                          "A_G_alleles",
                          "A_G_filtered_sites_raw_counts_wide.txt.gz"))

hmat <- read_tsv(file.path("hyperedits",
                          "diffedits",
                          "A_G_filtered_sites_raw_counts_wide.txt.gz"))

# check same columns
all(colnames(mat) == colnames(hmat))

# next keep sites from hyperediting if not found in gatk pipeline (although the values are the same)
mat_sites <- mat$edit_site[mat$edit_site %in% hmat$edit_site]
hmat <- hmat[!hmat$edit_site %in% mat_sites, ]

mat <- bind_rows(mat, hmat)

mats <- map(regexes, 
            ~mat %>% 
              dplyr::select(edit_site, matches(.x)) %>% 
              as.data.frame() %>% 
              tibble::column_to_rownames("edit_site"))
names(mats) <- region_order
ref <- "A"
alt <- "G"
pdata_groups <- map2(mats, 
                     regexes, 
                     ~get_groups(.x, .y, ref, alt))

# exclude Ent samples as they represent an ambiguous state

map(pdata_groups,
    ~dplyr::filter(.x,
                   State == "LT" | State == "Ar") %>% 
      dplyr::pull(g_names)) -> libs_to_exclude

map2(mats, 
    libs_to_exclude,
    ~.x[!colnames(.x) %in% .y]) -> mats

map(pdata_groups,
    ~dplyr::filter(.x,
                   State != "LT" & State != "Ar")) -> pdata_groups
```

```{r anova}
# get list of sig sites
gatk_sites <- read_tsv("edits/A_G_filtered_fdr0.01_sites_annotation_kmeans.txt.gz")
gatk_sites <- dplyr::filter(gatk_sites, 
                            kmeans_cluster == 1) %>% 
  dplyr::pull(site) %>% 
  unique()

hyperedited_sites <- read_tsv("hyperedits/diffedits/A_G_filtered_sites_annotation_kmeans.txt.gz")
hyperedited_sites <- hyperedited_sites %>% 
  dplyr::filter(kmeans_cluster != 3, 
                FDR < 0.01) %>% 
    dplyr::pull(site) %>% 
  unique()

sig_sites <- c(hyperedited_sites,
                   gatk_sites) %>% unique()

map(mats, ~.x[rownames(.x) %in% sig_sites, ]) -> mats

```


```{r}
# extract library sizes from # of uniquely mapped reads
.dir <- file.path(data_dir, "star")
star_report_files <- dir(.dir, 
                         pattern = "*_2pass_Log.final.out",
                         recursive = T, full.names = T)

star_reports <- map(star_report_files, 
                    ~read_tsv(.x, col_names = F))

star_report_names <- basename(star_report_files) %>% 
  str_replace(., "_2pass_Log.final.out", "")

names(star_reports) <- star_report_names
lib_sizes <- map(star_reports, 
                 ~dplyr::filter(.x, 
                                X1 == "Uniquely mapped reads number |") %>% 
                   dplyr::select(X2))

lib_sizes <- bind_rows(lib_sizes, .id = "library") %>% 
  dplyr::rename(lib_size = X2) %>% 
  dplyr::mutate(lib_size = as.numeric(lib_size))


extract_lib_sizes <- function(matrix_dat, lib_sizes_df) {
  #get original mapped to exons count for selected libraries
  .names <- colnames(matrix_dat)
  .names <- str_replace(.names, "_counts.txt.gz_a$", "")
  .names <- str_replace(.names, "_counts.txt.gz_b$", "")
  .names <- data_frame(original_name = .names)
  sizes <- inner_join(.names, lib_sizes_df, 
                      by = c("original_name" = "library"))
  res <- dplyr::select(sizes, lib_size) %>% unlist()
  names(res) <- sizes[["library"]]
  res
}

lib_sizes <- map(mats, 
                 ~extract_lib_sizes(.x, lib_sizes))
```

```{r build_design_matrices}
# design for testing explictly warm vs cold 
# winter-warm = IBA, Ent
# summer-warm = SA SpD

get_design_matrix <- function(pdata, ref_nt){

  #set up design matrix
  design <- model.matrix(~0 + animal_number + temp:allele_type, data = pdata)
  design <- design[,!grepl(paste0("allele_type", ref_nt), colnames(design))]
  design
}

design_matrices <- map(pdata_groups, ~get_design_matrix(.x, ref))

```


```{r initialize_edger_objs}

get_edger_objs <- function(matrix_dat, pdata, lib_size){
  dge <- DGEList(matrix_dat, group = pdata$temp, lib.size = lib_size)
  keep <- rowSums(getCounts(dge) > 5, na.rm = T) >= 4
  dge <- dge[keep, , keep.lib.sizes = TRUE]
  dge
}

edger_objs <- pmap(list(mats, pdata_groups, lib_sizes),
                   function(x, y, z) get_edger_objs(x, y, z))

```
### run normalization and DE analysis
```{r normalize_objs}
  
norm_edger_objs <- function(edger_obj, design_matrix) {
  #obtain normalization values for Reference alleles and propagate to ALT alleles
  dge_subset <- calcNormFactors(edger_obj[, 1:20, keep.lib.sizes = TRUE])
  edger_obj$samples$norm.factors <- rep(dge_subset$samples$norm.factors, 2)
  
  edger_obj <- estimateDisp(edger_obj, design_matrix)
  edger_obj
}

edger_objs <- map2(edger_objs, 
                   design_matrices, 
                   ~norm_edger_objs(.x, .y))
```

```{r set_up_glms}
get_topTags <- function(dge, design, alt_nt){
  
  fit <- glmFit(dge, design)
 
    #set up contrasts to test for cold versus warm
  colnames(design) <- str_replace(colnames(design), ":", ".")
  

  .cons <- "tempwinter_warm.allele_typeG-tempsummer_warm.allele_typeG"
  
  .contrasts <- makeContrasts(
                        contrasts = .cons,
                        levels = design)
  
  con <- glmLRT(fit, contrast = .contrasts)
  
  con <- topTags(con, n = nrow(con))
  
  con <- as.data.frame(con)
  
  tibble::rownames_to_column(con, "site")
}

de_results <- map2(edger_objs, design_matrices, 
                   ~get_topTags(.x, .y, alt))

```


```{r save_objects}
system("mkdir -p edits/winter-warm_v_summer-warm_edits")

walk2(de_results,
     names(de_results),
     ~write_gztsv(.x,
              file.path("edits",
                        "winter-warm_v_summer-warm_edits",
                        paste0(.y,
                        "_deresults.tsv"))))

cpms <- map(edger_objs, 
             ~cpm(.x, normalized.lib.sizes = T))
names(cpms) <- c("Medulla",
                     "Hypothalamus",
                     "Forebrain")

walk2(cpms, names(cpms), 
     ~as.data.frame(.x) %>% 
       tibble::rownames_to_column("site") %>% 
       write_gztsv(., file.path("edits",
                           "winter-warm_v_summer-warm_edits",
                           paste0(.y,
                           "_normalized_cpms.txt"))))

```

```{r read_objs, fig.height = 8, fig.width = 14 }

de_dat <- map(region_order, 
    ~read_tsv(file.path("edits",
                        "winter-warm_v_summer-warm_edits",
                        paste0(.x,
                        "_deresults.tsv.gz"))))

map(de_dat,
    ~dplyr::mutate(.x, significant = ifelse(FDR < 0.01,
                                            "significant",
                                            "not_significant")) %>% 
      dplyr::group_by(significant) %>% 
      dplyr::mutate(n_sites = n()) %>% 
      ungroup() %>% 
      dplyr::mutate(significant = factor(significant,
                                         levels = c("significant", "not_significant")),
                    significant = ifelse(significant == "significant",
                                          paste0("significant editing sites\nFDR < 0.01 n = ", n_sites),
                                          paste0("not significant\nn = ", n_sites))) %>% 
  dplyr::arrange(significant)) -> de_dat

names(de_dat) <- region_order

n_sig_down <- map(de_dat, ~.x %>% 
                  dplyr::filter(FDR < 0.01,
                                logFC < 0) %>% 
                  dplyr::pull() %>% length)
n_sig_up <- map(de_dat, ~.x %>% 
                  dplyr::filter(FDR < 0.01,
                                logFC > 0) %>% 
                  dplyr::pull() %>% length)

plts <- pmap(
  list(
    de_dat, 
    names(de_dat),
    n_sig_down,
    n_sig_up),
  function(x, y, down, up){
    
    y <- ifelse(y == "Forebrain",
                "Cerebrum",
                y)
    
    ggplot(x, aes(logCPM, logFC)) + 
    geom_point(aes(color = significant),
               alpha = 0.5,
               size = 0.25) +
    scale_color_brewer(palette = "Set1", 
                       name = "",
                       breaks = x$significant[nrow(x)]) +
    labs(title = y, 
         y = "Log2 Fold Change (Winter warm vs. Spring/summer warm)") +
    ylim(-17, 17) +
    annotate("text",
             x = 1.5, y = 14, label = paste0("increased frequency\nn = ", up)) +
    annotate("text",
             x = 1.5, y = -14, label = paste0("decreased frequency\nn = ", down)) +
    theme(
      axis.line.x = element_line(size = 1),
      axis.line.y = element_line(size = 1),
      legend.position = "none",
      legend.text = element_text(size = 12)
      )
  }
)


plt <- plot_grid(plotlist = plts, 
                 nrow = 1, 
                 align = "hv")

plt
save_plot("edits/winter-warm_v_summer-warm_edits/ma_plots.jpeg", 
          plt,
          device = "jpeg",
          dpi = 600,
          base_width = 10)

save_plot("edits/winter-warm_v_summer-warm_edits/ma_plots.pdf", 
          plt,
          base_width = 10)
```




```{r sig_sites_by_anova}

pdata_file <- file.path(docs_dir, "BrainRegionRNAseqLibMetadata.txt")
pdata <- readr::read_tsv(pdata_file)
  
pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)

pdata <- mutate(pdata, 
                abbrev_sample = str_split(sample, "_", simplify = T)[, 1])

pdata <- mutate(pdata,
                  temp = ifelse(str_detect(State,
                                           "IBA|Ent"),
                                "winter-warm",
                                ifelse(str_detect(State,
                                           "SA|SpD"),
                                "summer-warm", "cold")))
    

hyper_annots <- read_tsv(file.path("hyperedits", "diffedits", "A_G_filtered_sites_annotation_kmeans.txt.gz")) %>% 
  dplyr::filter(FDR < 0.01) %>% 
  dplyr::select(site, kmeans_cluster) %>% 
  unique()

gatk_annots <- read_tsv("edits/A_G_filtered_fdr0.01_sites_annotation_kmeans.txt.gz") %>% 
  dplyr::select(site, kmeans_cluster) %>% 
  unique()

list(hyper_annots,
     gatk_annots) -> annots

names(annots) <- c("hyper", "gatk")
annots <- bind_rows(annots, .id = "type")

de_dat_annot <- map(de_dat, 
                    ~left_join(.x, annots, by = "site"))
de_dat_annot <- bind_rows(de_dat_annot, .id = "region")

de_dat_annot %>% 
  dplyr::filter(!is.na(kmeans_cluster)) %>% 
  dplyr::select(site, logFC, FDR, type, kmeans_cluster) %>% 
  group_by(site) %>% 
  dplyr::arrange(type, .by_group = T) %>% # gatk > hyper
  summarize(logFC = max(logFC), 
            FDR = min(FDR), 
            type = dplyr::first(type), 
            kmeans_cluster = dplyr::first(kmeans_cluster)) -> grouped_dat

grouped_dat <- mutate(grouped_dat, significant = ifelse(FDR < 0.01,
                                         ifelse(logFC > 0,
                                                "significantly increased",
                                                "significantly decreased"),
                                         "not significant"),
                            annot_type = ifelse(type == "gatk",
                                                paste0("GATK Group ", kmeans_cluster),
                                                paste0("Hyperedited Group ", kmeans_cluster))) 

grouped_dat %>% 
  group_by(annot_type) %>% 
  mutate(grp_total = n()) %>% 
  group_by(annot_type, significant) %>% 
  summarize(prop = n(), 
            prop = prop / unique(grp_total)) -> summarized_grouped_dat

plt <- ggplot(summarized_grouped_dat, aes(annot_type, prop)) +
  geom_bar(aes(fill = significant), stat = "identity") +
  scale_y_continuous(labels = percent) +
  scale_fill_brewer(palette = "Greys",
                    name = "") +
  ylab("Percent of sites significant\nin cold vs. warm states") +
  guides(fill = guide_legend(nrow = 3)) +
  xlab("")  +
  theme(
    axis.text.x = element_text(angle = 90,
                               hjust = 1,
                               vjust = 0.5),
    legend.position = "top"
    
  )


#plt
#save_plot("edits/winter-warm_v_summer-warm_edits/prop_sig.pdf", plt, base_width = 3, base_height = 4.5)

grouped_dat %>% 
  dplyr::filter(FDR < 0.01) %>% 
  dplyr::pull(site) %>% 
  unique() -> sig_seasonal_sites


```

```{r hmaps, fig.width = 10, fig.height = 6}
# regenerate heatmap matrices

dat <- read_tsv(file.path("edits", "variant_allele_counts_by_strand", 
                             "A_G_alleles", "A_G_filtered_sites_proportions.txt.gz"))

dat_hyper <- read_tsv(file.path("hyperedits", "diffedits", "A_G_filtered_sites_proportions.txt.gz"))

dat_hyper <- dat_hyper %>% 
  anti_join(dat, by = "edit_site")
dat <- bind_rows(dat, dat_hyper)

dat <- as.data.frame(dat)
rownames(dat) <- dat[, 1]
dat <- dat[, -1]

# convert to percentages
dat <- dat * 100

pdata_file <- file.path(docs_dir, "BrainRegionRNAseqLibMetadata.txt")
 # make_groups
g_names <- colnames(dat)
animal_number <- str_extract(g_names, "[MHB][0-9]+")
  
 # allele_type <- rep(c("A", "G"), each = 30)
groups <- data_frame(g_names, animal_number)
  
pdata <- read_tsv(pdata_file)
  
pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)
  
pdata <- mutate(pdata,
                region = ifelse(region == "Forebrain",
                                "Cerebrum",
                                region))
pdata <- mutate(pdata, 
                  abbrev_sample = str_split(sample, "_", simplify = T)[, 1])
  
pdata <- inner_join(groups, 
                      pdata, 
                      by = c("animal_number" = "abbrev_sample")) %>%
    dplyr::select(g_names, animal_number, State, region)

pdata_heatmap <- data_frame( "names" = colnames(dat)) %>%
      inner_join(., pdata, by = c("names" = "g_names")) %>%
      dplyr::select(names, State, region) %>% as.data.frame() %>% 
      arrange(State)

reordered_cols <- pdata_heatmap$names %>% unlist()

dat <- dat[, reordered_cols]

pdata_heatmap <- dplyr::rename(pdata_heatmap, 
                               "Sample Groups" = State,
                               "Brain Region" = region)


ha2 <- HeatmapAnnotation(df = pdata_heatmap[, c("Sample Groups", "Brain Region")], 
                                      col = list(
                                        "Brain Region" = region_cols,
                                        "Sample Groups" = state_cols
                                                 ),
                           annotation_legend_param = list(
                             "Sample Groups" = list(title = "Sample Groups",
                                          at = state_order, 
                                          labels = state_order, nrow = 1),
                             "Brain Region" = list(title = "Brain Region", nrow = 3)), 
                           gap = unit(3, "mm"),
                           show_annotation_name = TRUE,
                           annotation_name_side = "left",
                           annotation_name_offset = unit(1, "mm"))

sig_dat_sites <- dat[rownames(dat) %in% sig_seasonal_sites, ]

n_sites <- nrow(sig_dat_sites)

ht <- Heatmap(sig_dat_sites,
      #        col = circlize::colorRamp2(c(0, 1), c("white", "red")),
          column_title = paste0("Editing sites\n", 
                                         "IBA + Ent  vs. SA + SpD",
                                         "\n n = ",
                                        n_sites),
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          top_annotation = ha2,
           row_title = "Editing Sites",
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")



sig_dat_sites_rotated <- t(sig_dat_sites)

ha2_row <- rowAnnotation(df = pdata_heatmap[, c("Sample Groups", "Brain Region")], 
                                      col = list(
                                        "Brain Region" = region_cols,
                                        "Sample Groups" = state_cols
                                                 ),
                           annotation_legend_param = list(
                             "Sample Groups" = list(
                                          at = state_order, labels = state_order)),
                       gap = unit(2, "mm"),
                           show_annotation_name = TRUE,
                           annotation_name_side = "top",
                           annotation_name_offset = unit(1, "mm"))


n_sites <- ncol(sig_dat_sites_rotated)

ht <- Heatmap(sig_dat_sites_rotated,
      #        col = circlize::colorRamp2(c(0, 1), c("white", "red")),
          column_title = paste0("Seasonally variable editing sites \n(n = ", formatC(n_sites, big.mark = ","), ")"),
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          show_row_dend = T,
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ha2_row + ht, annotation_legend_side = "left", gap = unit(1, "mm"), 
     heatmap_legend_side = "left", row_dend_side = "left", padding = unit(c(4, 8, 4, 4), "mm"))

pdf("edits/editing_sites_with_a_seasonal_component.pdf", width = 14, height = 8)
draw(ha2_row + ht, annotation_legend_side = "left", gap = unit(1, "mm"), 
     heatmap_legend_side = "left", row_dend_side = "left", padding = unit(c(4, 8, 4, 4), "mm"))
dev.off()

file.copy("edits/editing_sites_with_a_seasonal_component.pdf",
          figs_dir[5], overwrite = T)
```


```{r index}

raw_dat <- read_tsv(file.path("edits", "variant_allele_counts_by_strand","A_G_alleles/A_G_all_sites_raw_counts.txt.gz"))
raw_dat <- filter(raw_dat, edit_site %in% rownames(sig_dat_sites))

count_data <- raw_dat %>% 
  dplyr::select(-c(`#chrom`:ref)) %>% 
  gather(key = "library", value = "counts", -edit_site, -base)


count_data <- count_data %>% 
  left_join(pdata, by = c("library" = "g_names"))

# set order
count_data <- mutate(count_data,
                     State = factor(State, levels = state_order))

count_data <- count_data %>% 
  spread(base, counts)

index_data <- group_by(count_data, State, region) %>% 
  summarize(editing_index = sum(G) / (sum(A) + sum(G))) %>% 
  ungroup() 
  

ggplot(index_data, aes(State, editing_index)) +
  geom_point(aes(color = region)) +
  geom_smooth(method = "loess") +
  scale_colour_manual(values = region_cols) +
  ylab("Editing Index") +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(size = 12, 
                                   colour = state_cols),
        legend.position = "top",
        legend.title = element_blank())
```


```{r site_annotations}
gatk_sites <- read_tsv("edits/A_G_filtered_fdr0.01_sites_annotation_kmeans.txt.gz")
gatk_sites <- dplyr::filter(gatk_sites, 
                            kmeans_cluster == 1) %>% 
  dplyr::select(-EFF,
                Region:Disruption,
                Nucleotide:AminoAcid,
                GENE_ID:GENE_NAME, 
                kmeans_cluster) %>% 
    dplyr::rename(EFF = EFF_ensembl) %>% 
  unique()


hyperedited_sites <- read_tsv("hyperedits/diffedits/A_G_filtered_sites_annotation_kmeans.txt.gz")
hyperedited_sites <- hyperedited_sites %>% 
  dplyr::filter(kmeans_cluster != 3, 
                FDR < 0.01) %>% 
     dplyr::select(-EFF,
                Region:Disruption,
                Nucleotide:AminoAcid,
                GENE_ID:GENE_NAME, 
                kmeans_cluster) %>% 
  anti_join(gatk_sites, by = "site")

all_sig_site_annotation <- bind_rows(gatk_sites, hyperedited_sites)

all_sig_site_annotation %>%
  dplyr::filter(site %in% rownames(sig_dat_sites)) %>%
  dplyr::select(site, GENE_NAME) %>% 
  unique() 

seasonal_site_annotation <- all_sig_site_annotation %>%
  dplyr::filter(site %in% rownames(sig_dat_sites))

DT::datatable(seasonal_site_annotation)

write_tsv(seasonal_site_annotation, "edits/winter-warm_v_summer-warm_edits/sig_site_annotation.txt")

seasonal_site_annotation %>% 
  dplyr::select(site, GENE_NAME, EFF) %>% 
  unique() %>% 
  count(EFF)
```

## Plot editing frequencies for impactful sites

```{r plot_frequencies, fig.height = 24, fig.width = 12}
long_prop_dat <- read_tsv(file.path("edits", "variant_allele_counts_by_strand",
                                    "A_G_alleles", "A_G_filtered_sites_proportions_tidy.txt.gz"))

plot_sites <- function(prop_dat, 
                       toptag_dat, 
                       stat_region = "Medulla", 
                       region_to_plot = "all", 
                       sites_to_plot = c("GRIA2"),
                       save = T,
                       title = NULL, ...) {
  
    col_palette <- c(
    SA = rgb(255, 0, 0, maxColorValue=255),
    IBA = rgb(67, 205, 128, maxColorValue=255),
    Ent = rgb(155, 48, 255, maxColorValue=255),
    LT = rgb(25, 25, 112, maxColorValue=255),
    Ar = rgb(0, 0, 255, maxColorValue=255),
    SpD = rgb(255, 165, 0, maxColorValue=255) 
  )

  state_order = c(
    "SA", "IBA", "Ent", "LT", "Ar", "SpD"
  )

  tt_dat_med <- filter_(toptag_dat, 
                        lazyeval::interp(quote(Region == x), x = stat_region)) %>% 
    filter(site %in% sites_to_plot) %>%  
    dplyr::select(site) %>% 
    unlist()
  
  tt_dat_med <- filter(toptag_dat, site %in% tt_dat_med)
  
  if (!region_to_plot == "all"){
    prop_dat <- prop_dat %>% filter_(lazyeval::interp(quote(Region == x), 
                                                      x = region_to_plot))
  }
  
  plot_dat <- inner_join(prop_dat, 
                         tt_dat_med, 
                         by = c("site", "Region"))
  
  gene_labels <- plot_dat %>% 
    dplyr::select(GeneName) %>% 
    mutate(GeneName = str_split(GeneName, ",", simplify = T) %>% .[, 1]) %>% 
    unlist()
  
  names(gene_labels) <- plot_dat %>% 
    dplyr::select(site) %>% 
    unlist()
  
  
  plot_dat <- mutate(plot_dat, 
                State = factor(State, 
                        levels = state_order),
                Region = ifelse(Region == "Forebrain",
                                "Cerebrum",
                                Region),
                proportion = 100 * proportion)
    
  annot <- group_by(plot_dat, site, Region) %>% 
    summarize(FDR = paste0("FDR = ", signif(min(FDR), 3)))

  p <- ggplot(plot_dat, aes(State, proportion)) +
    #geom_boxplot(aes(fill = State), outlier.shape = NA) +
    geom_jitter(width = 0.25,
              aes(color = State)) +
    stat_summary(fun.y = "mean", 
               fun.ymin = "mean", 
               fun.ymax= "mean",
               size = 0.3,
               width = 0.75,
               geom = "crossbar") +
    facet_grid(Region ~ site, labeller = labeller(site = gene_labels)) + 
    scale_color_manual(values = col_palette, guide = guide_legend(nrow = 1)) +
    geom_text(data = annot, aes(x = 3.5, y = 95, label = FDR ), size = 3) + 
    ylab("Editing Frequency (%)") + 
    guides(fill=guide_legend(nrow=2,byrow=TRUE)) +
    theme(legend.position = "top",
          axis.text.x = element_blank(),
          strip.text = element_text(size = 9)
          )
  if(save){
    if(is.null(title)) {
      save_plot(paste0("edits/", gene_labels[1], "_boxplot.pdf"), p, ...)
    } else {
      save_plot(title, p, ...) 
    }
  }
  p
}

site_annotation <- read_tsv(file.path("edits", 
                                      "variant_allele_counts_by_strand", 
                                "A_G_alleles",
                                "A_G_filtered_sites_annotation.txt.gz"))
                        
impactful <- all_sig_site_annotation %>% 
  dplyr::filter(str_detect(IMPACT, "HIGH|MODERATE")) %>% 
  dplyr::select(site, GENE_NAME) %>% unique() 

impactful_sites <- impactful$site
names(impactful_sites) <- impactful$GENE_NAME

plts <- map(impactful_sites, ~plot_sites(prop_dat = long_prop_dat, 
                   toptag_dat = site_annotation,
                   sites_to_plot = .x))
plt <- plot_grid(plotlist = plts, ncol = 3)
plt
save_plot("edits/impactful_variants_editing_freq.pdf", plt,base_width= 12, base_height = 24)
```



## Heatmap of recoding events


```{r impactful_heatmap, fig.height = 10, fig.width = 8}

impactful_or_syn <- all_sig_site_annotation %>% 
  dplyr::filter(str_detect(IMPACT, "HIGH|MODERATE") | str_detect(EFF, "synonymous")) %>% 
  dplyr::select(site, GENE_NAME, AminoAcid) %>% 
  unique() %>% 
  mutate(AminoAcid = str_replace_all(AminoAcid, fixed("p."), "") %>% str_replace_all(",+$", "")) %>% 
  mutate(AminoAcid = ifelse(GENE_NAME == "ZCCHC8", 
                            "splice site",
                            AminoAcid),
         GENE_NAME = ifelse(GENE_NAME == "ENSSTOG00000022071",
                            "RLF",
                            GENE_NAME),
         GENE_NAME = ifelse(GENE_NAME == "ENSSTOG00000009473",
                            "ZNF483",
                            GENE_NAME),
         AminoAcid = str_split(AminoAcid, fixed(",")) %>% map_chr(., dplyr::first),
         GENE_NAME = ifelse(str_detect(GENE_NAME, "ENSSTOG00000016721"),
                            "RTL1",
                            GENE_NAME),
         AminoAcid = ifelse(AminoAcid == "Arg235Gly", 
                            "Arg378Gly",
                            AminoAcid)) %>% 
  mutate(hmap_name = str_c(GENE_NAME, AminoAcid, sep = " "))


impact_dat_sites <- dat[rownames(dat) %in% impactful_or_syn$site, ]

n_sites <- nrow(impact_dat_sites)

# make column order match state_order
left_join(data_frame(`Sample Groups` = state_order), 
          pdata_heatmap, 
          by = "Sample Groups") %>% 
  dplyr::pull(names) -> col_names_for_hmap

new_row_ids <- left_join(data_frame(site = rownames(impact_dat_sites)),
                         impactful_or_syn) %>% 
  dplyr::pull(hmap_name)

rownames(impact_dat_sites) <- new_row_ids

ht <- Heatmap(impact_dat_sites,
          col = c("white", "red"),
          column_title = paste0("CDS and impactful editing sites (n = ",
                                        n_sites, ")"),
          show_row_names = TRUE,
          show_row_dend = FALSE,
          show_column_names = FALSE,
          cluster_rows = T,
          cluster_columns = F,
          column_order = col_names_for_hmap,
          top_annotation = ha2,
          heatmap_legend_param = list(
            title = "Percent\nEditing (%)")
         )

draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left")

pdf("edits/all_cds_impactful_sites.pdf")
draw(ht, annotation_legend_side = "top", heatmap_legend_side = "left", padding = unit(c(4, 12, 4, 4), "mm"))
dev.off()

file.copy("edits/all_cds_impactful_sites.pdf",
          figs_dir[5], overwrite = T)
```