---
title: "Fastq quality checks"
author: "Kent Riemondy RBI"
date: "8/2/2017"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Check QC metrics for Fastq samples

## Read length

```{r, message=F, warning = F}
library(dplyr)
library(purrr)
library(readr)
library(stringr)
library(RColorBrewer)
library(cowplot)
source("globals.R") # defines paths for relevant directories
```

## Read alignments

```{r uniquely_mapped_reads_percentage, message=F, warning=F}
star_alignments <- list.files(file.path(data_dir, "star"), 
                              pattern = "[0-9]_2pass_Log.final.out",
                              recursive = T, full.names = T)

qc_dat <- suppressWarnings(map(star_alignments, ~readr::read_tsv(.x, col_names = F)))
names(qc_dat) <- star_alignments

qc_dat <- bind_rows(qc_dat, .id = "file")

qc_dat <- mutate(qc_dat, 
                 region = dirname(file) %>% 
  str_split(., "/", simplify = F) %>% 
  map_chr(., ~.x[8]))

uniq_reads_percent <- filter(qc_dat, X1 == "Uniquely mapped reads % |") %>% 
  mutate(X2 = as.numeric(str_replace(X2, "%", "")))

uniq_reads <- filter(qc_dat, X1 == "Uniquely mapped reads number |") %>% 
  mutate(X2 = as.numeric(X2))

uniq_reads <- mutate(uniq_reads, 
                     region = factor(region,
                                     levels = c("brainrest",
                                                "hypothalamus",
                                                "medulla"),
                                     labels = c("Cerebrum",
                                                "Hypothalamus",
                                                "Medulla")))

uniq_reads_percent <- mutate(uniq_reads_percent, 
                     region = factor(region,
                                     levels = c("brainrest",
                                                "hypothalamus",
                                                "medulla"),
                                     labels = c("Cerebrum",
                                                "Hypothalamus",
                                                "Medulla")))

a <- ggplot(uniq_reads, aes(file, X2)) + 
  geom_bar(stat = "identity", aes(fill = region)) +
  ylab("Number of Uniquely Mapped Reads") +
  scale_fill_manual(values = region_cols) + 
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top"
  )
a

mean(uniq_reads_percent$X2)
sd(uniq_reads_percent$X2)


raw_reads <- filter(qc_dat, X1 == "Number of input reads |") %>% 
  mutate(X2 = as.numeric(X2))

mean(raw_reads$X2)
sd(raw_reads$X2)

raw_reads$Unique <- filter(qc_dat, X1 == "Uniquely mapped reads number |") %>% 
  mutate(X2 = as.numeric(X2)) %>% dplyr::pull(X2)

raw_reads$Multi <- filter(qc_dat, X1 == "Number of reads mapped to multiple loci |") %>% 
  mutate(X2 = as.numeric(X2)) %>% dplyr::pull(X2)

mutate(raw_reads, 
       unmapped = X2 - Unique - Multi) -> raw_reads
```

Mean alignment rate = `r mean(uniq_reads_percent$X2)`
Sd alignment rate = `r sd(uniq_reads_percent$X2)`


Mean raw reads = `r mean(raw_reads$X2)`
Sd raw reads = `r sd(raw_reads$X2)`

Mean unmapped reads = `r mean(raw_reads$unmapped)`
Sd unmapped reads = `r sd(raw_reads$unmapped)`

```{r uniquely_mapped_reads, fig.width = 8}
b <- ggplot(uniq_reads_percent, aes(file, X2)) + 
  geom_bar(stat = "identity", aes(fill = region)) +
  ylab("Percent Uniquely Mapped Reads") +
  scale_fill_manual(values = region_cols) + 
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top"
  )
b

plt <- plot_grid(a, b, labels = "AUTO")
plt
save_plot("mrna/qc_alignments.pdf", plt, base_width = 8)
file.copy("mrna/qc_alignments.pdf", sfigs_dir[1], overwrite = T)
```

### write out table

```{r}
qc_dat_summary <- spread(qc_dat, X1, X2) 

colnames(qc_dat_summary) <- colnames(qc_dat_summary) %>% 
  str_replace(., " \\|", "") 

qc_dat_summary <- qc_dat_summary  %>% 
  dplyr::select(file, 
         region,
         `Number of input reads`,
         `Uniquely mapped reads %`,
         `% of reads mapped to multiple loci`)

qc_dat_summary <- qc_dat_summary %>% 
  mutate(fastq_id = str_split(file, "\\/", simplify = T) %>% .[, 9],
         sample_id = str_split(fastq_id, "_", simplify = T) %>% 
           .[, 1] %>% 
           str_replace("^[0-9]+-", ""),
         fastq_id = str_replace(fastq_id, "_2pass_Log.final.out", "")) %>% 
  dplyr::select(-file)

## get phenodata
pdata <- read_tsv(file.path(docs_dir, "BrainRegionRNAseqLibMetadata.txt"))
pdata <- gather(pdata, region, 
                sample, Forebrain, Hypothalamus, Medulla)
pdata <- mutate(pdata, 
                animal_number = str_split(sample, "_", simplify = T)[, 1])

pdata <- dplyr::filter(pdata,
                       !is.na(sample))

simple_pdata <- dplyr::select(pdata, animal_number, State, sacTb) %>% unique()


qc_dat_summary <- left_join(qc_dat_summary,
          simple_pdata, by = c("sample_id" = "animal_number")) %>% 
  mutate(region = ifelse(region == "brainrest",
                         "cerebrum",
                         region),
         sequencing_read_length = ifelse(region == "cerebrum",
                                         126,
                                         151))

## add in mapping between GEO title and fastq name

geo_ids <- tibble::tribble(
  ~sample_title, ~fastq_id,
       "Forebrain_SA_rep1",    "B30_2_GCCAAT_L001_R1_001",
      "Forebrain_SA_rep2",   "B31_8_AGTTCC_L001_R1_001",
      "Forebrain_SA_rep3",  "B32_14_CCGTCC_L002_R1_001",
      "Forebrain_SA_rep4",  "B33_20_ATTCCT_L002_R1_001",
      "Forebrain_SA_rep5",  "B34_26_GCCAAT_L003_R1_001",
      "Forebrain_Ar_rep1",   "B53_6_GATCAG_L001_R1_001",
      "Forebrain_Ar_rep2",  "B54_12_TGACCA_L002_R1_001",
     "Forebrain_Ent_rep1",   "B55_4_GTGAAA_L001_R1_001",
      "Forebrain_Ar_rep3",  "B57_18_GAGTGG_L002_R1_001",
      "Forebrain_Ar_rep4",  "B58_24_GGCTAC_L003_R1_001",
     "Forebrain_Ent_rep2",  "B59_10_GTCCGC_L001_R1_001",
      "Forebrain_Ar_rep5",  "B61_30_GATCAG_L003_R1_001",
     "Forebrain_Ent_rep3",  "B62_16_GTTTCG_L002_R1_001",
     "Forebrain_Ent_rep4",  "B64_22_ACTTGA_L003_R1_001",
      "Forebrain_LT_rep1",   "B65_5_TTAGGC_L001_R1_001",
     "Forebrain_Ent_rep5",  "B66_28_GTGAAA_L003_R1_001",
      "Forebrain_LT_rep2",  "B67_11_CGATGT_L002_R1_001",
      "Forebrain_LT_rep3",  "B68_17_CGTACG_L002_R1_001",
     "Forebrain_IBA_rep1",  "B69_15_GTGGCC_L002_R1_001",
     "Forebrain_IBA_rep2",  "B70_21_ATCACG_L003_R1_001",
      "Forebrain_LT_rep4",  "B72_23_TAGCTT_L003_R1_001",
      "Forebrain_LT_rep5",  "B74_29_TTAGGC_L003_R1_001",
     "Forebrain_SpD_rep1",  "B77_13_CAGATC_L002_R1_001",
     "Forebrain_IBA_rep3",  "B78_27_CTTGTA_L003_R1_001",
     "Forebrain_SpD_rep2",  "B79_19_ACTGAT_L002_R1_001",
     "Forebrain_SpD_rep3",  "B80_25_ACAGTG_L003_R1_001",
     "Forebrain_SpD_rep4",   "B83_1_ACAGTG_L001_R1_001",
     "Forebrain_SpD_rep5",   "B84_7_AGTCAA_L001_R1_001",
     "Forebrain_IBA_rep4",  "B153_3_CTTGTA_L001_R1_001",
     "Forebrain_IBA_rep5",  "B163_9_ATGTCA_L001_R1_001",
   "Hypothalamus_SA_rep1",     "35-H29_S39_L005_R1_001",
   "Hypothalamus_SA_rep2",     "41-H30_S45_L005_R1_001",
   "Hypothalamus_SA_rep3",     "47-H31_S60_L008_R1_001",
   "Hypothalamus_SA_rep4",     "53-H33_S66_L008_R1_001",
   "Hypothalamus_SA_rep5",     "59-H34_S72_L008_R1_001",
   "Hypothalamus_Ar_rep1",     "31-H53_S35_L005_R1_001",
   "Hypothalamus_Ar_rep2",     "37-H54_S41_L005_R1_001",
   "Hypothalamus_Ar_rep3",     "43-H57_S47_L005_R1_001",
  "Hypothalamus_Ent_rep1",     "32-H59_S36_L005_R1_001",
   "Hypothalamus_Ar_rep4",     "49-H60_S62_L008_R1_001",
   "Hypothalamus_Ar_rep5",     "55-H61_S68_L008_R1_001",
  "Hypothalamus_Ent_rep2",     "38-H62_S42_L005_R1_001",
  "Hypothalamus_Ent_rep3",     "44-H64_S48_L005_R1_001",
   "Hypothalamus_LT_rep1",     "34-H65_S38_L005_R1_001",
  "Hypothalamus_Ent_rep4",     "50-H66_S63_L008_R1_001",
   "Hypothalamus_LT_rep2",     "40-H67_S44_L005_R1_001",
   "Hypothalamus_LT_rep3",     "46-H68_S59_L008_R1_001",
  "Hypothalamus_IBA_rep1",     "33-H69_S37_L005_R1_001",
  "Hypothalamus_IBA_rep2",     "39-H70_S43_L005_R1_001",
   "Hypothalamus_LT_rep4",     "52-H72_S65_L008_R1_001",
   "Hypothalamus_LT_rep5",     "58-H74_S71_L008_R1_001",
  "Hypothalamus_SpD_rep1",     "48-H77_S61_L008_R1_001",
  "Hypothalamus_IBA_rep3",     "45-H78_S49_L005_R1_001",
  "Hypothalamus_SpD_rep2",     "54-H79_S67_L008_R1_001",
  "Hypothalamus_SpD_rep3",     "60-H80_S73_L008_R1_001",
  "Hypothalamus_SpD_rep4",     "36-H83_S40_L005_R1_001",
  "Hypothalamus_SpD_rep5",     "42-H84_S46_L005_R1_001",
  "Hypothalamus_IBA_rep4",    "51-H153_S64_L008_R1_001",
  "Hypothalamus_IBA_rep5",    "57-H163_S70_L008_R1_001",
  "Hypothalamus_Ent_rep5",    "56-H164_S69_L008_R1_001",
        "Medulla_SA_rep1",       "5-M30_S5_L001_R1_001",
        "Medulla_SA_rep2",     "11-M31_S11_L001_R1_001",
        "Medulla_SA_rep3",     "17-M32_S21_L003_R1_001",
        "Medulla_SA_rep4",     "23-M33_S27_L003_R1_001",
        "Medulla_SA_rep5",     "29-M34_S33_L003_R1_001",
        "Medulla_Ar_rep1",       "1-M53_S1_L001_R1_001",
        "Medulla_Ar_rep2",       "7-M54_S7_L001_R1_001",
       "Medulla_Ent_rep1",       "2-M55_S2_L001_R1_001",
        "Medulla_Ar_rep3",     "13-M57_S13_L001_R1_001",
        "Medulla_Ar_rep4",     "19-M58_S23_L003_R1_001",
       "Medulla_Ent_rep2",       "8-M59_S8_L001_R1_001",
        "Medulla_Ar_rep5",     "25-M61_S29_L003_R1_001",
       "Medulla_Ent_rep3",     "14-M62_S14_L001_R1_001",
        "Medulla_LT_rep1",       "4-M65_S4_L001_R1_001",
       "Medulla_Ent_rep4",     "20-M66_S24_L003_R1_001",
        "Medulla_LT_rep2",     "10-M67_S10_L001_R1_001",
        "Medulla_LT_rep3",     "16-M68_S20_L003_R1_001",
       "Medulla_IBA_rep1",       "3-M69_S3_L001_R1_001",
       "Medulla_IBA_rep2",       "9-M70_S9_L001_R1_001",
        "Medulla_LT_rep4",     "22-M72_S26_L003_R1_001",
        "Medulla_LT_rep5",     "28-M74_S32_L003_R1_001",
       "Medulla_SpD_rep1",     "18-M77_S22_L003_R1_001",
       "Medulla_IBA_rep3",     "15-M78_S15_L001_R1_001",
       "Medulla_SpD_rep2",     "24-M79_S28_L003_R1_001",
       "Medulla_SpD_rep3",     "30-M80_S34_L003_R1_001",
       "Medulla_SpD_rep4",       "6-M83_S6_L001_R1_001",
       "Medulla_SpD_rep5",     "12-M84_S12_L001_R1_001",
       "Medulla_IBA_rep4",    "21-M153_S25_L003_R1_001",
       "Medulla_IBA_rep5",    "27-M163_S31_L003_R1_001",
       "Medulla_Ent_rep5",    "26-M164_S30_L003_R1_001",
           "Neonate_head", "P1_H_33_CAGATC_L004_R1_001",
         "Neonate_middle", "P1_M_34_CCGTCC_L004_R1_001",
           "Neonate_tail", "P1_T_35_GTGGCC_L004_R1_001",
            "Testes_rep1", "T77D_32_TGACCA_L004_R1_001",
            "Testes_rep2", "T77O_31_CGATGT_L004_R1_001"
  )

geo_ids <- mutate(geo_ids, fastq_id = str_replace(fastq_id, "_R1_001", ""))

qc_dat_summary <- left_join(qc_dat_summary, 
                            geo_ids, 
                            by = "fastq_id")

qc_dat_summary <- dplyr::select(qc_dat_summary,
                fastq_id, sample_title, 
                sample_id, region, State, sacTb, everything()) %>% 
  dplyr::rename(`Number of reads after quality and adapter trimming` = `Number of input reads`)
openxlsx::write.xlsx(qc_dat_summary, "mrna/qc_summary_statistics.xlsx")
```

## metadat formatting for GEO

```{r geo}
# get phenotype data
pdata <- read_tsv(file.path(docs_dir, "BrainRegionRNAseqLibMetadata.txt"))
pdata <- gather(pdata, region, 
                sample, Forebrain, Hypothalamus, Medulla)
pdata <- mutate(pdata, 
                animal_number = str_split(sample, "_", simplify = T)[, 1])

pdata <- dplyr::filter(pdata,
                       !is.na(sample))

#get list of fastq filenames
fqfiles <- read_tsv("edits/variant_allele_counts_by_strand/A_G_alleles/A_G_all_sites_raw_counts.txt.gz", n_max = 1) %>% 
  colnames(.) %>% 
  str_subset(., "counts") %>% 
  str_replace(., "_counts.txt.gz", "_R1_001.fastq.gz")

r2 <- str_replace(fqfiles, "R1", "R2")

fqfiles <- data_frame(r1 = fqfiles, r2 = r2)

fqfiles <- mutate(fqfiles, 
                  animal_number = str_match(r1, "[MHB][0-9]+[^_]")[, 1])

metadata <- inner_join(pdata, fqfiles)

metadata <- metadata %>% 
  group_by(region, State) %>% 
  mutate(rep = row_number(),
         rep = paste0("rep", rep),
         sample_name = paste(region, State, rep, sep = "_"))

metadata <- dplyr::select(metadata,
                          -c(1:3), -SacDate, -rep, -sample, -animal_number)

metadata <- mutate(metadata,
                   organism = "Ictidomys tridecemlineatus",
                   molecule = "polyA_RNA",
                   sample_id = sample_name
                   )
# format to match xlsx for geo submission
metadata <- dplyr::select(metadata, 
                          sample_id, sample_name,
                          region, organism,
                          Animal, State,
                          Sex, sacTb, molecule,
                          everything())

metadata

write_tsv(metadata, file.path(docs_dir, "geo_formatted_metadata.txt"))
```


# percent of alignments to introns/exons

```{r, message = F, warning=F}
files <- dir(file.path(data_dir,
                         "star",
                         "align_stat"),
               recursive = T,
             full.names = T)

stats <- map(files, read_tsv, col_names = F) 

stats <- map(stats, 
             ~separate(.x, X2, c("id", "count"), sep = " "))
stats <- map(stats, ~.x[, -1])
stats <- bind_rows(stats, .id = "lib")

stats <- stats %>% 
  group_by(lib) %>%
  mutate(type = ifelse(str_detect(id, "exon"),
                       "Exonic",
                       ifelse(str_detect(id, fixed(".")), 
                       "Unannotated",
                       "Intron")))

stats %>% 
  group_by(lib, type) %>% 
  summarize(reads = sum(as.numeric(count))) %>% 
  group_by(lib) %>% 
  mutate(percent = 100 * (reads / sum(reads))) -> stats

stats %>% 
  ungroup() %>% 
  mutate(lib = as.numeric(lib),
         annotation = ifelse(lib <= 90,
                             "stringtie + taco",
                             "ensembl")) %>%
  dplyr::select(-lib) -> stats

p <- ggplot(stats, aes(type, percent)) +
  geom_boxplot(aes(fill = annotation), 
               coef = 10,
               lwd = 0.25) +
  scale_fill_brewer(palette = "Set1",
                    name = "Transcriptome") +
  labs(y = "Percent of uniquely aligned\nreads overlapping annotation",
       x = "")

save_plot("mrna/percent_exonic_intronic.pdf", p, base_aspect_ratio = c(1.4, 1))
file.copy("mrna/percent_exonic_intronic.pdf",
          sfigs_dir[1], overwrite = T)
stats %>% 
  group_by(annotation, type) %>% 
  summarize(mean_percent = mean(percent),
            sd_percent = sd(percent))

stats %>% 
  group_by(annotation) %>% 
  summarize(total_unique_alignments = sum(reads))
```
