---
title: "RNA-Seq kmeans"
author: "Kent Riemondy RBI"
date: "7/31/2017"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```
## Goals

- Classify into clusters of gene expression

## Cluster data using kmeans and compare with Random Forest

```{r}
# from results/2017-07-31/kmeans.Rmd
library(dplyr)
library(valr)
library(purrr)
library(readr)
library(stringr)
library(tidyr)
library(edgeR)
library(RColorBrewer)
library(ComplexHeatmap)
library(flexclust)
library(cluster)
library(cowplot)
library(eulerr)
library(randomForest)
source("globals.R")

state_cols <-  c(
  SA = rgb(255, 0, 0, maxColorValue=255),
  IBA = rgb(67, 205, 128, maxColorValue=255),
  Ent = rgb(155, 48, 255, maxColorValue=255),
  LT = rgb(25, 25, 112, maxColorValue=255),
  Ar = rgb(0, 0, 255, maxColorValue=255),
  SpD = rgb(255, 165, 0, maxColorValue=255) 
)

state_order = c(
  "SA", "IBA", "Ent", "LT", "Ar", "SpD"
)

#colors for region
region_cols <- c(
  Medulla = "#4DAF4A",
  Hypothalamus = "#377EB8",
  Forebrain =  "#E41A1C"
)
# region order
region_order <- c("Medulla", "Hypothalamus", "Forebrain")

color_fxn <- colorRampPalette(brewer.pal(9, "Spectral"))

get_sig_heatmap <- function(vector_of_sites, 
                            data_matrix_to_subset, 
                            title = "", 
                            pdata_mat = pdata, ...){  
  all_sig_sites <- data_matrix_to_subset[rownames(data_matrix_to_subset) %in% vector_of_sites, ]

  n_sites = nrow(all_sig_sites)
  
  pdata_mat <- data_frame( "names" = colnames(data_matrix_to_subset)) %>%
      inner_join(. , pdata_mat, by = c("names" = "sample")) %>%
      dplyr::select(names, State, region) %>% 
      unique() %>% 
      as.data.frame() %>% 
      arrange(State)
  
  reordered_cols <- pdata_mat$names %>% unlist()

  all_sig_sites <- all_sig_sites[, reordered_cols]

  ha2 <- HeatmapAnnotation(df = pdata_mat[c('State', 'region')], 
                                      col = list(
                                        region = region_cols,
                                          State = state_cols
                                                 ),
                           annotation_legend_param = list(
                             State = list(
                               title = "Sampling\nGroup",
                               at = state_order, 
                               labels = state_order),
                             region = list(
                               title = "Region"
                             )))
  ht <- Heatmap(all_sig_sites, 
          show_column_names = FALSE,
          top_annotation = ha2,
          row_title = "Transcripts",
          column_title = paste0(title, "\n", n_sites, " transcripts"),
          heatmap_legend_param = list(
            title = expression("Log"[2]*" CPM")),
          ...)
  
  draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")
  
}

write_gztsv <- function(df, name, ...){
  # write output as gzipped, supply name without .gz
  if(str_detect(name, ".gz$")){
    uncompressed_name <- str_replace(name, ".gz$", "")
  } else {
    uncompressed_name <- name
  }
  write_tsv(df, uncompressed_name, ...)
  R.utils::gzip(uncompressed_name, remove = T, overwrite = T)
}

```

### specify sample information
```{r}
# get sample information and counts
counts <- read_tsv(file.path(data_dir, "featurecounts", "count_summary.tsv"),
                  skip = 1)
counts <- as.data.frame(counts)
rownames(counts) <- counts$Geneid
counts <- counts[-c(1:6)] #drop gene interval information

fq_names <- colnames(counts)
fq_names <- basename(fq_names)
fq_names <- str_split(fq_names, "_2pass", simplify = T)[, 1]
colnames(counts) <- fq_names

```


```{r}
# get pdata about animal state
pdata <- read_tsv(file.path(docs_dir, "BrainRegionRNAseqLibMetadata.txt"))
pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)
  
pdata <- mutate(pdata, 
                abbrev_sample = str_split(sample, "_", simplify = T)[, 1]) %>% 
  dplyr::rename(long_sample = sample)

animal_number <- str_extract(fq_names, "[MHB][0-9]+")

fq_dat <- data_frame(sample = fq_names, 
                     animal_number)

pdata <- inner_join(fq_dat, 
                    pdata, 
                    by = c("animal_number" = "abbrev_sample"))

simplified_pdata <- dplyr::select(pdata, sample, State, region) %>% 
  unique() %>% 
  mutate(State = factor(State, 
                        levels = c("SA", "IBA", "Ent", "LT", "Ar", "SpD")))

```

## Examine sample relationships with random forest

```{r}

# read in data, ignorning col names because there are only nfields - 1 colnames
dat <- map(region_order,
           ~read_tsv(paste0("mrna/", .x, "_normalized_counts.txt"),
                     col_names = F, 
                     skip = 1))

# get column names as vector and add in a transcript field at start
cnames <- map(region_order,
           ~read_tsv(paste0("mrna/", .x, "_normalized_counts.txt"),
                     col_names = F,
                     n_max = 1) %>% 
             .[1, ] %>% 
             unlist() %>% 
             c("transcript", .))
names(dat) <- region_order

# assign column names...
for (i in seq_along(dat)) {
  colnames(dat[[i]]) <- cnames[[i]]
}  

# convert to standard dataframe for clustering/heatmaps/etc.
dat <- map(dat, ~as.data.frame(.x))

# fix rownames
for (i in seq_along(dat)) {
  rownames(dat[[i]]) <- dat[[i]][, 1]
  dat[[i]][, 1] <- NULL
}

```

```{r fig.width = 12, fig.height = 4}
if(!file.exists("mrna/randomforest.rds")){
# extract out top 5000 variable genes for classification
  gene_variance <- map(dat, ~matrixStats::rowVars(as.matrix(.x)))
  variable_gene_idxs <- map(gene_variance, ~order(.x, decreasing = TRUE)[1:5000])
  vargenes <- map2(dat, variable_gene_idxs, ~.x[.y, ])
  
  # run randomForest
  rfs <- map(vargenes, ~randomForest(t(.x), ntree = 20000))
  saveRDS(rfs, "mrna/randomforest.rds")
}
  rfs <- readRDS("mrna/randomforest.rds")
  #plot MDS of dissimilarity values
  mds_dat <- map(rfs, 
                 ~stats::cmdscale(1 - .x$proximity, eig = TRUE, k = 2))
  
  mds_point_dat <- map(mds_dat, 
                       ~as.data.frame(.x$points) %>% 
    mutate(sample = rownames(.)) %>% 
    dplyr::rename(dim1 = V1,
           dim2 = V2) %>% 
    tbl_df())
  
  mds_point_dat <- map(mds_point_dat, ~inner_join(.x, 
                              pdata, by = "sample"))
  names(mds_point_dat) <- region_order
  
  plts <- map2(mds_point_dat, names(mds_point_dat),
               ~ggplot(.x, aes(dim1, dim2, color = State)) +
                  geom_text(aes(label = State)) + 
                  scale_color_manual(values = state_cols) +
                  ggtitle(.y) + 
                  theme(legend.pos = "none"))
  
  plt <- plot_grid(plotlist = plts, align = "hv", 
            nrow = 1)
  plt
  save_plot(file.path(figs_dir[1], "RF_mds_plots.pdf"), plt, base_height = 4, base_width = 12)

```

## Classify transcripts

Group transcripts into classes describing their patterns during hibernation.

### pull out significant sites


```{r}
# extract significant sites
so_dat <- map(region_order,
           ~read.table(paste0("mrna/", .x, "_lrt_results.txt"),
                     header = T, sep = "\t", stringsAsFactors = F) %>% 
  as_data_frame())

so_dat <- map(so_dat,
              ~dplyr::filter(.x, FDR < 0.01))

names(so_dat) <- region_order

sig_trans <- map(so_dat, ~.x$denovo_id)
```

### run kmeans clustering
```{r clustering}
sig_mats <- map2(dat, sig_trans, ~.x[rownames(.x) %in% .y, ] )

mean_center <- function(mat) {
  #log + 1 and mean center
  mat <- log2(mat + 1)
  mat <- mat - rowMeans(mat)
  mat
}

sig_mats <- map(sig_mats, mean_center)

set.seed(20170721)
e_km <- map(sig_mats, 
            ~flexclust::kcca(.x, k = 5, 
                        control = list(initcent = "kmeanspp")))
km_clusters <- map(e_km, ~.x@cluster)

```

```{r, eval = T}

elbow_calc <- function(mat, k){
  # compute sum of squares for k = 1
  wss <- (nrow(mat)-1)*sum(apply(mat,2,var))
  for( i in 2:k) {
    clust <- flexclust::kcca(mat, i, control = list(initcent = "kmeanspp"))
    wss[i] <- info(clust, "distsum")
  }
  plot(1:k, wss, type="b", xlab="Number of Clusters",
     ylab="Within groups sum of squares")
}

map(sig_mats, ~elbow_calc(.x, 10))
```

```{r fig.width= 8, fig.height = 12}
pwalk(list(
  sig_trans,
  sig_mats,
  km_clusters),
  function(x, y, z) 
     get_sig_heatmap(x,
                y,
                show_row_dend = T,
                show_row_names = F,
                split = z))

pwalk(list(
  sig_trans,
  sig_mats,
  km_clusters,
  names(sig_mats)),
  function(x, y, z, outname) {
     pdf(paste0(figs_dir[1], "/", outname, "_edgerheatmap.pdf"), 
         height = 8, width = 6, useDingbats = F)
     get_sig_heatmap(x,
                y,
                show_row_dend = T,
                show_row_names = F,
                split = z)
     dev.off()
     }
  )

```

### summarize expression patterns per cluster
```{r editpattern, fig.width= 8 ,fig.height = 12}
# tidy
cluster_data <- map(sig_mats, ~mutate(as.data.frame(.x),
                                      transcript = rownames(.x)))
                    
cluster_data <- map(cluster_data,
                    ~as_data_frame(.x) %>% 
  tbl_df() %>% 
  gather(key = "library", value = "transformed_editing_freqs", -transcript) )

names(cluster_data) <- region_order

# get pdata
clustered_annotated_data <- map(cluster_data, ~inner_join(.x, unique(pdata), 
                                       by = c("library" = "sample")))

# get cluster id
cid <- map(km_clusters, ~data_frame(transcript = names(.x),
                  cluster = .x))

clustered_annotated_data <- map2(clustered_annotated_data, 
                                 cid,
                                 ~inner_join(.x ,.y,  
                                       by = "transcript"))
names(clustered_annotated_data) <- region_order
clustered_annotated_df <- bind_rows(clustered_annotated_data, .id = "Region")

# set order
clustered_annotated_df <- mutate(clustered_annotated_df,
                                   State = factor(State, levels = state_order))

# boxplot
#ggplot(clustered_annotated_df, aes(State, transformed_editing_freqs)) +
#  geom_boxplot(aes(fill = State)) +
#  facet_wrap(Region~cluster) +
#  scale_fill_manual(values = state_cols) +
#  ylab("Editing Frequency, \nmean centered and scaled")

# summarize and show as mean +/- sd
clustered_annotated_df <- group_by(clustered_annotated_df, cluster, Region) %>% 
  mutate(n = length(unique(transcript)))

summary_dat <- group_by(clustered_annotated_df, State, cluster, Region, n) %>% 
  summarize(mean_editing = mean(transformed_editing_freqs, na.rm = T),
            sd_editing = sd(transformed_editing_freqs, na.rm = T))

tx_counts <- clustered_annotated_df %>% 
  dplyr::select(State, Region, cluster, n) %>% 
  mutate(n = paste0("n = ", n), 
         x = 2, y = 2) %>% 
  unique()

ggplot(summary_dat, aes(State, mean_editing)) +
  geom_point(aes(color = as.factor(cluster))) +
  geom_line(aes(group = as.factor(cluster),
                                    colour = as.factor(cluster))) +
  geom_errorbar(aes(ymin = mean_editing - sd_editing, 
                    ymax = mean_editing + sd_editing), width = .1) +
  geom_text(aes(x = x, y = y, group = cluster, label = n), 
            data = tx_counts,
            check_overlap = T) +
  facet_grid(cluster~Region) +
  scale_colour_brewer("Cluster Id", palette = "Dark2") +
  ylab("Transcript abundance TPM \nmean centered") 
```

### reorder clustering to match patterns across regions
```{r, fig.width = 8}

reordered_classes = list(
  Medulla = as.character(c(1:5)),
  Hypothalamus = as.character(c(1:5)),
  Forebrain = as.character(c(1:5))
)
# group into classes
reordered_classes_order = list(
  Medulla = as.character(c(4, 5, 3, 1, 2)),
  Hypothalamus = as.character(c(4, 3, 5, 1, 2)),
  Forebrain = as.character(c(1, 3, 2, 5, 4))
)

for (i in seq_along(reordered_classes)) {
  names(reordered_classes[[i]]) <- reordered_classes_order[[i]]
}

x <- map2(clustered_annotated_data, 
     reordered_classes, 
     ~mutate(.x, 
             cluster = as.character(cluster),
             cluster = unname(.y[cluster])))

clustered_annotated_df <- bind_rows(x, .id = "Region")

# set order
clustered_annotated_df <- mutate(clustered_annotated_df,
                                   State = factor(State, levels = state_order))

# boxplot
#ggplot(clustered_annotated_df, aes(State, transformed_editing_freqs)) +
#  geom_boxplot(aes(fill = State)) +
#  facet_wrap(Region~cluster) +
#  scale_fill_manual(values = state_cols) +
#  ylab("Editing Frequency, \nmean centered and scaled")

# summarize and show as mean +/- sd
clustered_annotated_df <- group_by(clustered_annotated_df, cluster, Region) %>% 
  mutate(n = length(unique(transcript)))

summary_dat <- group_by(clustered_annotated_df, State, cluster, Region, n) %>% 
  summarize(mean_editing = mean(transformed_editing_freqs, na.rm = T),
            sd_editing = sd(transformed_editing_freqs, na.rm = T))

tx_counts <- clustered_annotated_df %>% 
  dplyr::select(State, Region, cluster, n) %>% 
  ungroup() %>% 
  mutate(n = paste0("n = ", n), 
         x = 1.75, y = 1.5,
         cluster = paste0("cluster ", cluster)) %>% 
  unique()

summary_dat <- ungroup(summary_dat) %>% 
  mutate(color = state_cols[State],
         cluster = paste0("cluster ", cluster)) %>% 
  group_by(cluster)
 
ggplot(summary_dat, aes(State, mean_editing)) +
  geom_point(aes(color = cluster)) +
  geom_line(aes(group = cluster, colour = cluster)) +
  geom_errorbar(aes(ymin = mean_editing - sd_editing, 
                    ymax = mean_editing + sd_editing), width = .1) +
  geom_text(aes(x = x, y = y, group = cluster, label = n), 
            data = tx_counts,
            check_overlap = T) +
  facet_grid(cluster~Region) +
  scale_color_manual(values = c("black", "black", "black", "grey", "grey"),
                     name = "Kmeans cluster Id") +
  ylim(-1, 1.75) +
  ylab(expression("CPM mean centered log"[2])) +
  theme(axis.text.x = element_text(colour = state_cols),
        plot.margin = unit(c(4, 10, 4, 4), "mm"),
        legend.pos ="none")
ggsave(file.path(figs_dir[1], "cluster_summary_edger.pdf"), width = 7, height = 6)

```


## shared transcripts in each class

```{r}
clustered_annotated_df %>% 
  dplyr::select(Region, transcript, cluster) %>% 
  unique() -> tx_per_class

split_by_cluster <- map(c(1, 2, 3, 4, 5), 
    ~filter(tx_per_class, cluster == .x))

split_by_cluster <- map(split_by_cluster,
      ~split(.x, .x$Region))

only_txs <- map(split_by_cluster,  
                ~map(.x, ~pull(.x, transcript)))

shared_txs <- map(only_txs, ~unlist(.x) %>% unique())

lgl_df <- list()
for (i in seq_along(only_txs)){
  lgl_df[[i]] <- map_df(only_txs[[i]], ~shared_txs[[i]] %in% .x)
}
names(lgl_df) <- paste("Class", c(1:5))

lgl_df <- map(lgl_df, as.data.frame)

map2(lgl_df, names(lgl_df), 
    ~plot(euler(.x),
                  counts = T,
                  fill = region_cols,
                  fill_opacity = 0.55,
                  main = .y))


map2(lgl_df, names(lgl_df), 
    function(x, y){
      pdf(paste0("mrna/", str_replace(y, " ", "_"), "_venn.pdf"), useDingbats = F)
      p <- plot(euler(x),
                  #counts = T,
                  fill = region_cols,
                  fill_opacity = 0.55,
                  counts = list(cex = 4), 
                  lwd = 6,
                  main = NULL,
                labels = NULL)
      print(p)
      dev.off()
    }
)

tmp <- bind_rows(lgl_df, .id = "class")
tmp$class <- factor(tmp$class)

plot(euler(tmp[, c(2:4)], by = tmp[, 1]),
#    counts = T,
#    fill = region_cols,
    fill_opacity = 0.55,
#  auto.key = TRUE,
    layout = c(1, 5))
```

## Add class to output files

```{r}

clustered_annotated_df %>% 
  dplyr::select(Region, transcript, cluster) %>% 
  unique() %>% 
  write_gztsv("mrna/transcript_kmeans_class.txt")
```
```{r, eval = T}

# extract out transcripts shared in each class in each region and plot all together
class_txs <- list()
for (i in seq_along(only_txs)){
  class_txs[[i]] <- map_df(only_txs[[i]], ~shared_txs[[i]] %in% .x ) 
  class_txs[[i]]$tid <- shared_txs[[i]]
}

class_txs <- map(class_txs, ~.x[rowSums(.x[, c(1:3)]) == 3, ]) 
shared_class_txs <- map(class_txs, ~pull(.x, tid)) %>% 
  unlist()

names(class_txs) <- c(1:5)

class_txs_df <- bind_rows(class_txs, .id = "cluster")
cluster_ids <- pull(class_txs_df, cluster)  %>% as.numeric()
names(cluster_ids) <- pull(class_txs_df, tid)

# generate single large matrix
shared_mats <- map(dat, ~.x[rownames(.x) %in% shared_class_txs, ] )
map(shared_mats, ~tibble::rownames_to_column(.x, "genes")) -> tmp
Reduce(function(x, y) inner_join(x, y, by = "genes"), tmp) -> shared_mat
rownames(shared_mat) <- shared_mat$genes
shared_mat[, 1] <- NULL

shared_mat <- mean_center(shared_mat)
cluster_ids <- cluster_ids[order(match(cluster_ids, rownames(shared_mat)))]
#pdf(paste0(outname, "_edgerheatmap.pdf"), height = 8, width = 6, useDingbats = F)

get_sig_heatmap(shared_class_txs,
                shared_mat,
                show_row_dend = T,
                show_row_names = F,
                split = cluster_ids)
```

```{r, eval = F}

edits <- read_tsv("../2017-04-17/variant_allele_counts_by_strand/A_G_alleles/A_G_filtered_sites_annotation.txt.gz")

sig_edits <- filter(edits, FDR < 0.10)

sig_edits$collapsed_GeneID <- str_split(sig_edits$GeneID, ",") %>%
    map(unique) %>% map(~.x[.x != ""]) %>%
    map_chr(~ifelse(length(.x) > 1, paste(.x, collapse = ","), .x))

sig_edits <- dplyr::filter(sig_edits, !str_detect(collapsed_GeneID, "-"))
  
sig_edits <- split(sig_edits, sig_edits$Region)

sig_edits <- sig_edits[c(2, 1, 3)]

sig_edits <- map(sig_edits, ~dplyr::select(.x ,collapsed_GeneID) %>% 
                   unlist())

# read in annotation data
annots <- read_tsv("../../dbases/ensembl85/Ictidomys_tridecemlineatus.spetri2.85.bed.gz", col_names = F)

geneid_mapping <- dplyr::select(annots, X7, X8, X10) %>% 
  unique()

colnames(geneid_mapping) <- c("geneId", "transcriptpId", "geneName")

# compare to editing sites

dat <- map(region_order,
      ~read_tsv(paste0("sleuth_results_", .x, ".txt")))

names(dat) <- region_order

sig_txns <- map(dat, 
  ~filter(.x, qval < 0.01) %>% 
    dplyr::select(ens_gene) %>% 
    unique() %>% 
    unlist())


list_of_exons_and_genes <- transpose(list(genes_with_editing = sig_edits, 
                                          DE_genes = sig_txns))

library(eulerr)
plot_prop_venn <- function(gene_lst, region){
  all_unique_sites <- unlist(gene_lst) %>% unique()
  lgl_df <- map_df(gene_lst, ~all_unique_sites %in% .x)
  fit <- euler(lgl_df)
  
  #pdf("exon_edits_venn_diagrams.pdf")
  #plt <- plot(fit, counts = TRUE, 
  #           key = list(space = "left", columns = 1), 
  #           fill = region_cols, 
  #           fill_opacity = 0.75,
  #           main = "Shared Significant Exons")
  #print(plt)
  #dev.off()
  
  plot(fit, counts = TRUE, 
        key = list(space = "left", columns = 1), 
        fill = region_cols, 
        fill_opacity = 0.75,
        main = region)
}

map2(list_of_exons_and_genes, 
     names(list_of_exons_and_genes), 
           function(x, y) {
             pdf(paste0(y, "_de_genes_edit_sites.pdf"))
             print(plot_prop_venn(x, region = y))
             dev.off()
             })

map2(list_of_exons_and_genes, 
     names(list_of_exons_and_genes), 
             ~plot_prop_venn(.x, region = .y))


```


## make heatmap of ncRNA

```{r read_annotation_dat, fid.width= 10, fig.height = 20}
annot_dat <- map(region_order,
    ~read_tsv(paste0("mrna/", .x, "_lrt_results.txt")))

names(annot_dat) <-region_order
annot_dat <- bind_rows(annot_dat, .id = "Region")
classes <- read_tsv("mrna/transcript_kmeans_class.txt.gz")

annot_dat <- left_join(annot_dat, classes, 
                       by = c("Region", "denovo_id" = "transcript"))
annot_dat %>% dplyr::filter(FDR < 0.01,
                            cluster == 5) -> tmp

tmp %>% 
  group_by(denovo_id, gene_name_ensembl) %>% 
  summarize(n = n()) %>% 
  dplyr::filter(n >=2,
                !is.na(gene_name_ensembl)) %>% 
  dplyr::pull(denovo_id) -> c5_genes

tmp %>% 
  group_by(denovo_id, gene_name_ensembl) %>% 
  summarize(n = n()) %>% 
  dplyr::filter(n >=2,
                !is.na(gene_name_ensembl)) %>% 
  dplyr::select(denovo_id, gene_name_ensembl) -> c5_gene_names

mutate(c5_gene_names, 
       new_id = ifelse(is.na(gene_name_ensembl),
                       denovo_id,
                       gene_name_ensembl)) -> new_ids

# read in data, ignorning col names because there are only nfields - 1 colnames
cpm_dat <- map(region_order,
           ~read_tsv(paste0("mrna/", .x, "_normalized_counts.txt"),
                     col_names = F, 
                     skip = 1))

# get column names as vector and add in a transcript field at start
cnames <- map(region_order,
           ~read_tsv(paste0("mrna/", .x, "_normalized_counts.txt"),
                     col_names = F,
                     n_max = 1) %>% 
             .[1, ] %>% 
             unlist() %>% 
             c("transcript", .))
names(cpm_dat) <- region_order

# assign column names...
for (i in seq_along(cpm_dat)) {
  colnames(cpm_dat[[i]]) <- cnames[[i]]
}  

# convert to standard dataframe for clustering/heatmaps/etc.
cpm_dat <- map(cpm_dat, ~as.data.frame(.x))
cpm_dat <- Reduce(function(x, y) inner_join(x, y, by = "transcript"), 
                  cpm_dat)
                  
# fix rownames

rownames(cpm_dat) <- cpm_dat[, 1]
cpm_dat[, 1] <- NULL

c5_exp <- cpm_dat[rownames(cpm_dat) %in% c5_genes, ]
data_frame(id = rownames(c5_exp)) %>% 
  inner_join(new_ids, by = c("id" = "denovo_id")) %>% 
  group_by(new_id) %>% 
  dplyr::mutate(n = row_number()) %>% 
  ungroup() %>% 
  dplyr::mutate(new_id = ifelse(n > 1,
                          paste0(new_id,"_",n),
                          new_id)) %>% 
  dplyr::pull(new_id) -> new_ids_matched

rownames(c5_exp) <- new_ids_matched
scaled_c5_exp <- t(scale(t(c5_exp)))

fq_names <- colnames(c5_exp)
pdata <- read_tsv(file.path(docs_dir, "BrainRegionRNAseqLibMetadata.txt"))
pdata <- gather(pdata, region, sample, Forebrain, Hypothalamus, Medulla)
  
pdata <- mutate(pdata, 
                abbrev_sample = str_split(sample, "_", simplify = T)[, 1]) %>% 
  dplyr::rename(long_sample = sample)

animal_number <- str_extract(fq_names, "[MHB][0-9]+")

fq_dat <- data_frame(sample = fq_names, 
                     animal_number)

pdata <- inner_join(fq_dat, 
                    pdata, 
                    by = c("animal_number" = "abbrev_sample"))

simplified_pdata <- dplyr::select(pdata, sample, State, region) %>% 
  unique() %>% 
  mutate(State = factor(State, 
                        levels = c("SA", "IBA", "Ent", "LT", "Ar", "SpD")))

 n_sites = nrow(scaled_c5_exp)
  
  pdata_mat <- data_frame( "names" = colnames(scaled_c5_exp)) %>%
      inner_join(. , pdata, by = c("names" = "sample")) %>%
      dplyr::select(names, State, region) %>% 
      unique() %>% 
      as.data.frame() %>% 
      arrange(State)
  
  reordered_cols <- pdata_mat$names %>% unlist()

  scaled_c5_exp <- scaled_c5_exp[, reordered_cols]

  ha2 <- HeatmapAnnotation(df = pdata_mat[c('State', 'region')], 
                                      col = list(
                                        region = region_cols,
                                          State = state_cols
                                                 ),
                           annotation_legend_param = list(
                             State = list(
                               title = "Sampling\nGroup",
                               at = state_order, 
                               labels = state_order),
                             region = list(
                               title = "Brain\nRegion"
                             )), 
                           gap = unit(2, "mm"),
                           show_annotation_name = TRUE,
                           annotation_name_offset = unit(1, "mm"))

  ht <- Heatmap(scaled_c5_exp, 
          show_column_names = FALSE,
          top_annotation = ha2,
          row_title = "Transcripts",
          column_title = paste0("Genes strongly\n enriched in Torpor"),
          heatmap_legend_param = list(
            title = expression("Z-score\nCPM")))
  
  draw(ht, annotation_legend_side = "left", heatmap_legend_side = "left")
  
pdf("mrna/class5_edgerheatmap.pdf", height = 6, width = 8, useDingbats = F)
draw(ht, annotation_legend_side = "left",
     gap = unit(1, "mm"), 
     heatmap_legend_side = "left", 
     row_dend_side = "left",
     padding = unit(c(4, 8, 4, 4), "mm"))
dev.off()
```
